---
title: "Quality Control"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcb_file: "data/YYYY-MM-DD/bcb.rda"
    min_umis: 1000
    max_umis: !r Inf
    min_genes: 500
    max_genes: !r Inf
    max_mito_ratio: 0.1
    min_novelty: 0.75
    min_cells_per_gene: 3
    output_dir: "."
---

```{r setup, cache=FALSE, message=FALSE, warning=FALSE}
# Last modified 2018-02-12
bcbioSingleCell::prepareSingleCellTemplate()
source("_setup.R")

# Directory paths
data_dir <- file.path(params$output_dir, "data", Sys.Date())

# Load bcbioSingleCell object
bcb_name <- load(params$bcb_file)
bcb <- get(bcb_name, inherits = FALSE)
stopifnot(is(bcb, "bcbioSingleCell"))
```

```{r header, child="_header.Rmd"}
```



```{r sample_metadata}
sampleMetadata(bcb)
```

[bcbio][] run data was imported from **`r metadata(bcb)$uploadDir`**.



# Quality control metrics {.tabset}

## Reads per cell

These are counts of how many reads are assigned to a given cellular barcode. The purpose of these histograms is that we should see one single large peak that represents cells that were encapsulated. If we see a strong shoulder, or a bimodal distribution of the cells, that can indicate a couple problems. It might be that there is free floating RNA, which happens when cells are dying. It could also be that there are a set of cells that failed for some reason. Finally, it could also be that there are biologically different types of cells, and one type is much smaller than the other. If this is the case we would expect to see less RNA being sequenced from the smaller cells.

```{r plot_reads_per_cell}
plotReadsPerCell(bcb, interestingGroups = "sampleName")
```


## Cell counts

```{r plot_cell_counts}
plotCellCounts(bcb)
```


## UMI counts per cell {.tabset}

```{r plot_umis_per_cell}
mdHeader("histogram", level = 3)
plotUMIsPerCell(
    bcb,
    geom = "histogram",
    min = min(params$min_umis)
)
    
mdHeader("violin", level = 3)
plotUMIsPerCell(
    bcb,
    geom = "violin",
    min = min(params$min_umis)
)

mdHeader("boxplot", level = 3)
plotUMIsPerCell(
    bcb,
    geom = "boxplot",
    min = min(params$min_umis)
)
```


## Genes detected per cell {.tabset}

Here by "detected", we mean genes with a non-zero count measurement per cell. Seeing gene detection in the range of `500`-`5000` is normal for most single-cell experiments.

```{r plot_genes_per_cell}
mdHeader("histogram", level = 3)
plotGenesPerCell(
    bcb,
    geom = "histogram",
    min = min(params$min_genes),
    max = max(params$max_genes)
)

mdHeader("violin", level = 3)
plotGenesPerCell(
    bcb,
    geom = "violin",
    min = min(params$min_genes),
    max = max(params$max_genes)
)

mdHeader("boxplot", level = 3)
plotGenesPerCell(
    bcb,
    geom = "boxplot",
    min = min(params$min_genes),
    max = max(params$max_genes)
)
```


## UMIs vs. genes detected

If we graph out the total number of UMI counts per cell vs. the genes detected per cell, we can assess whether there is a large population of low quality cells with low counts and/or gene detection.

```{r plot_umis_vs_genes}
plotUMIsVsGenes(bcb, interestingGroups = "sampleName")
```


## Mitochondrial counts ratio {.tabset}

We evaluate overall mitochondrial gene expression as a biomarker of cellular stress during sample preparation.

```{r plot_mito_ratio}
mdHeader("histogram", level = 3)
plotMitoRatio(
    bcb,
    geom = "histogram",
    max = max(params$max_mito_ratio)
)

mdHeader("violin", level = 3)
plotMitoRatio(
    bcb,
    geom = "violin",
    max = max(params$max_mito_ratio)
)

mdHeader("boxplot", level = 3)
plotMitoRatio(
    bcb,
    geom = "boxplot",
    max = max(params$max_mito_ratio)
)
```


## Novelty {.tabset}

Another way to QC the data is to look for less novelty, that is cells that have less genes detected per count than other cells. We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturated the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that we have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric.

```{r plot_novelty}
mdHeader("histogram", level = 3)
plotNovelty(
    bcb,
    geom = "histogram",
    min = min(params$min_novelty)
)

mdHeader("violin", level = 3)
plotNovelty(
    bcb,
    geom = "violin",
    min = min(params$min_novelty)
)

mdHeader("boxplot", level = 3)
plotNovelty(
    bcb,
    geom = "boxplot",
    min = min(params$min_novelty)
)
```



# Filter cells

```{r filter_cells}
bcb_filtered <- filterCells(
    bcb,
    minUMIs = params$min_umis,
    maxUMIs = params$max_umis,
    minGenes = params$min_genes,
    maxGenes = params$max_genes,
    maxMitoRatio = params$max_mito_ratio,
    minNovelty = params$min_novelty,
    minCellsPerGene = params$min_cells_per_gene)
assignAndSaveData(
    name = paste(bcb_name, "filtered", sep = "_"),
    object = bcb_filtered,
    dir = data_dir)
```

```{r plot_filtered_qc}
plotQC(
    bcb_filtered,
    geom = "violin",
    return = "markdown",
    headerLevel = 2)
```



```{r footer, child="_footer.Rmd"}
```
