<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Changelog" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="bcbiosinglecell-0-1-17-2018-07-09" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-17-2018-07-09" class="anchor"></a>bcbioSingleCell 0.1.17 (2018-07-09)</h1>
<div id="major-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes" class="anchor"></a>Major changes</h2>
<ul>
<li>
<code><a href="../reference/readCellRanger.html">readCellRanger()</a></code> now supports HDF5 data (.h5 files) or MatrixMarket Exchange Format explicitly, using the <code>format</code> argument. The function also now supports user-requested import of raw whitelisted cellular barcodes, using the <code>filtered</code> argument. By default, filtered counts from Cell Ranger are imported.</li>
<li>Removed support for zingeR in <code>diffExp()</code>. We will consider adding this functionality back in a future update, once the package is available on Bioconductor. For now, we’re recommending usage of zinbwave.</li>
<li>
<code>filterCells</code> nCells</li>
<li>Internal data has been renamed to snake_case format, for improved consistency. Note that <code>cell_cycle_markers</code> is used in place of <code>cellCycleMarkers</code>, and <code>cell_type_markers</code> in place of <code>cellTypeMarkers</code>. There’s no easy way to link these files as aliases, so this is a necessary breaking change.</li>
</ul>
</div>
<div id="minor-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes" class="anchor"></a>Minor changes</h2>
<ul>
<li>Improved R Markdown template defaults for quality control and clustering.</li>
<li>Now importing functions related to differential expression into the package NAMESPACE, including DESeq2, edgeR, and zinbwave.</li>
<li>Parallelization is no longer used to load samples, improving handling in an HPC cluster environment. We will consider using BiocParallel in a future update.</li>
<li>Now requiring [ggplot2][] v3.0. Switched to using <code>aes()</code> instead of <code>aes_string()</code> internally, using new tidyeval syntax.</li>
<li>Removed additional single-cell toolkit packages in Suggests, including scater and scran. This helps speed up build checks on Travis CI.</li>
<li>Reexported functions have been moved to the bcbioBase package.</li>
<li>Improved internal consistency between the <code><a href="../reference/bcbioSingleCell.html">bcbioSingleCell()</a></code> and <code><a href="../reference/readCellRanger.html">readCellRanger()</a></code> data import functions.</li>
</ul>
</div>
<div id="additional-notes" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-notes" class="anchor"></a>Additional notes</h2>
<ul>
<li>Package still works with Bioconductor 3.6 / R 3.4, but we will be migrating to Bioconductor 3.7 / R 3.5 when conda r-base is updated to 3.5.1.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-16-2018-06-20" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-16-2018-06-20" class="anchor"></a>bcbioSingleCell 0.1.16 (2018-06-20)</h1>
<div id="new-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions" class="anchor"></a>New functions</h2>
<ul>
<li>
<code>cellCountsPerCluster()</code> and <code>clusterCellCountsPerSample()</code>.</li>
</ul>
</div>
<div id="r-markdown-templates" class="section level2">
<h2 class="hasAnchor">
<a href="#r-markdown-templates" class="anchor"></a>R Markdown templates</h2>
<ul>
<li>Now recommending 1000 UMIs by default, and a maximum of 10% mitochondrial transcripts.</li>
<li>Seurat clustering now calculates multiple resolutions, suggesting 0.4, 0.8, and 1.2 by default. Dimensional reduction plots have been updated to support looping of these multiple resolutions.</li>
</ul>
</div>
<div id="minor-changes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-1" class="anchor"></a>Minor changes</h2>
<ul>
<li>`as(object, “SingleCellExperiment”) coercion now slots stashed metadata into SingleCellExperiment, if defined.</li>
<li>Simplified the internal code for <code><a href="../reference/sampleData.html">sampleData()</a></code>. <code>seurat</code> objects now share the same code as <code>SingleCellExperiment</code>, and return <code>NULL</code> if the sample data is not defined. The <code><a href="../reference/metrics.html">metrics()</a></code> function continues to slot empty metadata in <code>sampleID</code>, <code>sampleName</code>, and <code>interestingGroups()</code> if not defined.</li>
<li>
<code>SingleCellExperiment</code> is used as default method over <code>seurat</code> where applicable: <code>fetchData</code> family, <code>plotDimensionalReduction</code> family, <code>plotMarker</code> family, <code>plotFeature</code> family, <code>plotGene</code> family, and <code>plotCellTypesPerCluster()</code>. The internal code hasn’t changed, it just is defined primarily for <code>SingleCellExperiment</code>.</li>
<li>
<code>dimRed</code> argument has been renamed to <code>reduction</code>, where applicable.</li>
<li>
<code>topBarcodes()</code> can now return either a <code>data.frame</code> or <code>list</code>, containing the top barcodes grouped by sample.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-15-2018-06-13" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-15-2018-06-13" class="anchor"></a>bcbioSingleCell 0.1.15 (2018-06-13)</h1>
<div id="minor-changes-2" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-2" class="anchor"></a>Minor changes</h2>
<ul>
<li>Updated internal code to use <code>text</code> as primary argument in <code>markdownHeader()</code> calls.</li>
<li>Updated example datasets and unit tests to match.</li>
<li>Working on using pbmc4k as example dataset for vignette.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-14-2018-06-05" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-14-2018-06-05" class="anchor"></a>bcbioSingleCell 0.1.14 (2018-06-05)</h1>
<div id="major-changes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-1" class="anchor"></a>Major changes</h2>
<ul>
<li>Multiplexed <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> barcodes are now reformatted to be more readable and compatible with the <code><a href="../reference/mapCellsToSamples.html">mapCellsToSamples()</a></code> utility function. This change helps simplify the internal code for <code>cell2sample()</code>. For example in the pbmc4k dataset, the barcode IDs are sanitized from <code>TTTGGTTTCGCTAGCG-1</code> to “<code>pbmc4k_1_TTTGGTTTCGCTAGCG</code>”. The “<code>1</code>” here denotes 1 sample in the matrix, which is how <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> denotes multiplexed samples in a single counts matrix. Note that the “<code>-</code>” character is illegal in names, so we consistently sanitize barcodes to contain “<code>_</code>” instead. See <code>help("make.names")</code> for more information on syntactically valid names in <a href="https://www.r-project.org">R</a>.</li>
<li>
<code><a href="../reference/readCellRanger.html">readCellRanger()</a></code> no longer requires reference data defined by <code>refdataDir</code>, although this is still recommended.</li>
</ul>
</div>
<div id="minor-changes-3" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-3" class="anchor"></a>Minor changes</h2>
<ul>
<li>Resaved example datasets.</li>
<li>Switched the example <code>cellranger_small</code> and <code>seurat_small</code> datasets to the publicly available pbmc4k dataset from 10X Genomics. Here we’ve subset the top 500 cells and genes by abundance. We’ll use either the pbmc4k or pbmc8k dataset for the vignette in a future update.</li>
<li>
<code><a href="../reference/bcbioSingleCell.html">bcbioSingleCell()</a></code> and <code><a href="../reference/readCellRanger.html">readCellRanger()</a></code> functions now consistently default to not requiring <code>sampleMetadataFile</code>, which is now <code>NULL</code> by default. For <code><a href="../reference/bcbioSingleCell.html">bcbioSingleCell()</a></code>, if a custom sample metadata file is not provided, the function reads from the bcbio YAML metadata. For <code><a href="../reference/readCellRanger.html">readCellRanger()</a></code>, the function uses <code>minimalSampleData()</code> internally to return minimal metadata, containing <code>sampleName</code> and <code>description</code> columns.</li>
<li>Broke out internal <code>.sampleDirs()</code> function to <code><a href="../reference/bcbioSingleCell.html">bcbioSingleCell()</a></code> and <code><a href="../reference/readCellRanger.html">readCellRanger()</a></code> functions.</li>
<li>Improved <code>plotMarker</code> documentation examples to use mitochondrial genes.</li>
<li>Using <code>seurat_small</code> in place of <code><a href="http://www.rdocumentation.org/packages/Seurat/topics/pbmc_small">Seurat::pbmc_small</a></code> in working examples.</li>
</ul>
</div>
<div id="internal-code-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#internal-code-changes" class="anchor"></a>Internal code changes</h2>
<ul>
<li>Now consistently using <a href="https://dplyr.tidyverse.org">dplyr</a> for piped <code>data.frame</code> as much as possible, where applicable. Code is being updated to use <a href="https://dplyr.tidyverse.org/articles/programming.html">tidyeval</a>.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-13-2018-05-29" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-13-2018-05-29" class="anchor"></a>bcbioSingleCell 0.1.13 (2018-05-29)</h1>
<div id="minor-changes-4" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-4" class="anchor"></a>Minor changes</h2>
<ul>
<li>t-SNE and UMAP plot improvements.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-12-2018-05-25" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-12-2018-05-25" class="anchor"></a>bcbioSingleCell 0.1.12 (2018-05-25)</h1>
<div id="major-changes-2" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-2" class="anchor"></a>Major changes</h2>
<ul>
<li>No longer using automatic camel case sanitization for <code><a href="../reference/metrics.html">metrics()</a></code> or <code>fetchData</code> return column names.</li>
<li>Improved R Markdown clustering and marker templates to optionally support UMAP and dark mode in the YAML parameters.</li>
</ul>
</div>
<div id="minor-changes-5" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-5" class="anchor"></a>Minor changes</h2>
<ul>
<li>Using original <a href="http://satijalab.org/seurat">Seurat</a> mapping names for data: tSNE_1, tSNE_2, PC1, PC2, UMAP1, UMAP2.</li>
<li>Ensure transcript-level counts always have <code>stripTranscriptVersions()</code> command applied, to remove the Ensembl transcript versions if present.</li>
<li>No longer using labels (e.g. A, B, C, D) on <code>ggplot</code> grid return.</li>
<li>Note that “phase” has been renamed to “Phase” in the R Markdown clustering for cell-cycle regression PCA.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-11-2018-05-23" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-11-2018-05-23" class="anchor"></a>bcbioSingleCell 0.1.11 (2018-05-23)</h1>
<div id="new-functions-1" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions-1" class="anchor"></a>New functions</h2>
<ul>
<li>UMAP is now supported. This functionality is provided in: <code><a href="../reference/plotDimensionalReduction.html">plotUMAP()</a></code>, <code><a href="../reference/plotMarker.html">plotMarkerUMAP()</a></code>, and <code><a href="../reference/plotFeature.html">plotFeatureUMAP()</a></code>. Corresponding fetch functions, <code><a href="../reference/fetchData.html">fetchUMAPData()</a></code> and <code><a href="../reference/fetchData.html">fetchUMAPExpressionData()</a></code>, have also been added.</li>
<li>
<code><a href="../reference/plotGene.html">plotGene()</a></code>: Added <code>seurat</code> method support. If advanced customization of the plot is needed, use <code><a href="../reference/plotDot.html">plotDot()</a></code> or <code><a href="../reference/plotViolin.html">plotViolin()</a></code> instead, or refer to the <a href="http://satijalab.org/seurat">Seurat</a> documentation for alternates.</li>
</ul>
</div>
<div id="major-changes-3" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-3" class="anchor"></a>Major changes</h2>
<ul>
<li>Dimensional reduction and marker plots no longer use dark mode by default. The default color palette support for marker plots has been improved to consistently use viridis.</li>
<li>
<code>diffExp()</code>: improved internal code to work directly on <code>SingleCellExperiment</code>, removing the need to pass <code>design</code> and <code>group</code> parameters internally. Also added unit testing against <a href="https://doi.org/doi:10.18129/B9.bioc.zinbwave">zinbwave</a>, <a href="https://github.com/statOmics/zingeR">zingeR</a>, and [edgeR][] support. [DESeq2][] is supported but runs slowly.</li>
<li>Reworked <code><a href="../reference/plotFeature.html">plotFeature()</a></code> and <code>plotMarker()</code> family of functions. Improved the color palette support when <code>dark = FALSE</code>, now using a flipped viridis plasma color palette.</li>
<li>
<code><a href="../reference/aggregateReplicates.html">aggregateReplicates()</a></code> function has been reworked to return a <code>SingleCellExperiment</code> object instead of <code>bcbioSingleCell</code>. The v0.2.4 update of <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> behaves similarly with this generic.</li>
</ul>
</div>
<div id="minor-changes-6" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-6" class="anchor"></a>Minor changes</h2>
<ul>
<li>Reworked the internal handling of some <code>seurat</code> <code>SingleCellExperiment</code> method support, using <code><a href="../reference/coerce.html">as(x, "SingleCellExperiment")</a></code> internally, which uses the new <code><a href="http://www.rdocumentation.org/packages/Seurat/topics/Convert">Seurat::Convert()</a></code> function.</li>
<li>Made some previously deprecated functions now defunct: <code><a href="../reference/defunct.html">plotClusters()</a></code>, <code><a href="../reference/defunct.html">plotTSNEExpressionData()</a></code>, <code><a href="../reference/defunct.html">loadSingleCellRun()</a></code>, <code><a href="../reference/defunct.html">darkTheme()</a></code>, <code><a href="../reference/defunct.html">pcCutoff()</a></code>, <code><a href="../reference/defunct.html">quantileHeatmap()</a></code>, <code><a href="../reference/defunct.html">plotKnownMarkers()</a></code>, <code><a href="../reference/defunct.html">readMarkers()</a></code>, <code><a href="../reference/defunct.html">readMarkersFile()</a></code>.</li>
<li>Made <code><a href="../reference/defunct.html">plotFeatures()</a></code>, <code>plotMarker()</code>, and <code><a href="../reference/defunct.html">plotMarkers()</a></code> functions defunct.</li>
<li>
<code>plotPCElbow()</code> now returns a plot grid.</li>
<li>
<code>sanitizeMarkers()</code>: improved internal code for supported bcbio stashed metadata, including <code>rowRanges</code>.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-10-2018-05-19" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-10-2018-05-19" class="anchor"></a>bcbioSingleCell 0.1.10 (2018-05-19)</h1>
<div id="minor-changes-7" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-7" class="anchor"></a>Minor changes</h2>
<ul>
<li>
<code>plotCellTypesPerCluster()</code> is using <code>dark = TRUE</code> by default again.</li>
<li>Fixed <code>cell2sample()</code> handling for multiplexed Cell Ranger data loaded up with <code><a href="../reference/readCellRanger.html">readCellRanger()</a></code>. Need to use stashed <code>cell2sample</code> factor saved in <code>metadata()</code>, rather than attempting to calculate on the fly with <code><a href="../reference/mapCellsToSamples.html">mapCellsToSamples()</a></code>.</li>
<li>Updated <a href="https://travis-ci.org">Travis CI</a> build checks to include bioc-release on macOS.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-9-2018-05-18" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-9-2018-05-18" class="anchor"></a>bcbioSingleCell 0.1.9 (2018-05-18)</h1>
<div id="major-changes-4" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-4" class="anchor"></a>Major changes</h2>
<ul>
<li>No longer attempting to sanitize the rownames for <code>seurat</code> objects in coercion method. This helps maintain the gene symbol appearance in plotting functions for genes with hyphens in the names.</li>
</ul>
</div>
<div id="minor-changes-8" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-8" class="anchor"></a>Minor changes</h2>
<ul>
<li>Using <code><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/SerialParam-class">BiocParallel::SerialParam()</a></code> internally for zinbwave in <code>diffExp()</code>.</li>
<li>Simplified <code>cell2sample()</code> internal code to always use <code><a href="../reference/mapCellsToSamples.html">mapCellsToSamples()</a></code> instead of attempting to use a stashed <code>vector</code> inside <code>metadata()</code> for <code>SingleCellExperiment</code> method.</li>
<li>Removed internal <code>.applyFilterCutoffs()</code>, which is no longer necessary since this functionality is supported in the S4 subset method.</li>
<li>Simplified assert checks inside <code>fetchGene()</code> functions.</li>
<li>
<code>plotCellTypesPerCluster()</code>: revert back to <code>dark = TRUE</code> by default.</li>
<li>Consolidated <code>plotMarker</code> and <code>plotFeature</code> functions in the documentation.</li>
<li>
<code>sanitizeMarkers()</code>: Improved gene identifier matching.</li>
<li>
<code>topMarkers()</code> now defaults to <code>coding = FALSE</code> by default, since not all datasets will contain biotype information.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-8-2018-05-16" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-8-2018-05-16" class="anchor"></a>bcbioSingleCell 0.1.8 (2018-05-16)</h1>
<div id="minor-changes-9" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-9" class="anchor"></a>Minor changes</h2>
<ul>
<li>Initial <code><a href="../reference/updateObject.html">updateObject()</a></code> method support for <code>bcbioSingleCell</code> class.</li>
<li>Relaxed <code>validObject()</code> validity check to not require sample-level metadata in <code>colData()</code> yet.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-7-2018-05-15" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-7-2018-05-15" class="anchor"></a>bcbioSingleCell 0.1.7 (2018-05-15)</h1>
<div id="new-functions-2" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions-2" class="anchor"></a>New functions</h2>
<ul>
<li>Added support for Uniform Manifold Approximation and Projection (UMAP) with the <code><a href="../reference/plotDimensionalReduction.html">plotUMAP()</a></code> and <code><a href="../reference/fetchData.html">fetchUMAPData()</a></code> functions. These work similarly to the other <code><a href="../reference/plotDimensionalReduction.html">plotDimensionalReduction()</a></code> and <code><a href="../reference/fetchData.html">fetchData()</a></code> functions.</li>
</ul>
</div>
<div id="major-changes-5" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-5" class="anchor"></a>Major changes</h2>
<ul>
<li>Now adding sample-level metadata into <code>colData()</code> slot, for better downstream compatibility with other packages that work with <code>SingleCellExperiment</code> container class. Unique per-sample rows are still saved internally in the <code><a href="../reference/sampleData.html">sampleData()</a></code> slot.</li>
<li>Now recommending ECDF as the default geom for quality control plots, where applicable.</li>
<li>
<code>filterCells()</code> now supports <code>minUMIs = c("knee", "inflection")</code> for automatic filtering based on the cellular barcode ranks. Internally this is handled by <code><a href="http://www.rdocumentation.org/packages/DropletUtils/topics/barcodeRanks">DropletUtils::barcodeRanks()</a></code>.</li>
</ul>
</div>
<div id="minor-changes-10" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-10" class="anchor"></a>Minor changes</h2>
<ul>
<li>Attempting to re-enable <code>libgsl-dev</code> installation for <a href="https://doi.org/doi:10.18129/B9.bioc.zinbwave">zinbwave</a> on <a href="https://travis-ci.org">Travis CI</a>.</li>
<li>Suggesting <a href="https://doi.org/doi:10.18129/B9.bioc.BiocParallel">BiocParallel</a> for <a href="https://doi.org/doi:10.18129/B9.bioc.zinbwave">zinbwave</a> call in <code>diffExp()</code>.</li>
<li>Now importing <a href="http://satijalab.org/seurat">Seurat</a> functions into NAMESPACE.</li>
<li>Consolidated <code><a href="../reference/fetchData.html">fetchData()</a></code> functions in the documentation.</li>
<li>Consolidated <code><a href="../reference/plotDimensionalReduction.html">plotDimensionalReduction()</a></code> functions in the documentation.</li>
<li>Updated <code><a href="../reference/aggregateReplicates.html">aggregateReplicates()</a></code> internal code. This function again only supports aggregation of <code>bcbioSingleCell</code> objects that have been filtered using the <code>filterCells()</code> function.</li>
<li>Now using <code><a href="http://www.rdocumentation.org/packages/Seurat/topics/Convert">Seurat::Convert()</a></code> internally to coerce <code>seurat</code> class object to <code>SingleCellExperiment</code>, using <code><a href="../reference/coerce.html">as(seurat, "SingleCellExperiment")</a></code>. This utility function was added to <a href="http://satijalab.org/seurat">Seurat</a> v2.3.1.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-6-2018-05-09" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-6-2018-05-09" class="anchor"></a>bcbioSingleCell 0.1.6 (2018-05-09)</h1>
<div id="minor-changes-11" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-11" class="anchor"></a>Minor changes</h2>
<ul>
<li>Updated default QC R Markdown template.</li>
<li>Added trendline option to QC scatterplot functions.</li>
<li>Simplified internal handling of interestingGroups in <code><a href="../reference/plotQC.html">plotQC()</a></code>.</li>
<li>Using <code>readYAMLSampleData()</code> internally instead of defunct <code>sampleYAMLMetadata()</code>.</li>
<li>Added <code>sampleNames()</code> method support for seurat.</li>
<li>Now importing rmarkdown, sessioninfo, tidyverse for R Markdown reports, rather than suggesting. Similar update applied to bcbioRNASeq.</li>
<li>Suggesting scater and scran.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-5-2018-05-04" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-5-2018-05-04" class="anchor"></a>bcbioSingleCell 0.1.5 (2018-05-04)</h1>
<div id="major-changes-6" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-6" class="anchor"></a>Major changes</h2>
<ul>
<li>Overhauled inflection and knee point labeling support in <code>plotUMIsPerCell()</code>. Now uses the <code>point</code> argument and always labels per sample. Currently requires the <code>geom = "ecdf"</code> argument for labeling.</li>
<li>Updated default quality control template.</li>
<li>Added <code>plotBarcodeRanks()</code>.</li>
</ul>
</div>
<div id="minor-changes-12" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-12" class="anchor"></a>Minor changes</h2>
<ul>
<li>Added barcode rank support for <code>seurat</code> class objects.</li>
<li>QC plots now have titles by default, matching the conventions used in bcbioRNASeq.</li>
<li>Fixed y-axis scale for histogram geom in QC plots.</li>
<li>Prefiltering of very low quality barcodes with no UMIs or genes is now always applied. This helps avoid unwanted downstream errors with zero count barcodes.</li>
<li>Added boxplot geom support for `plotReadsPerCell().</li>
<li>
<code><a href="../reference/plotQC.html">plotQC()</a></code> geom argument is now more consistent across the paneled plots.</li>
<li>Fixed facet wrapping for aggregate samples in the QC plots.</li>
<li>Added <code>interestingGroups</code> support to <code>plotZerosVsDepth()</code>, matching the other QC functions.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-4-2018-04-30" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-4-2018-04-30" class="anchor"></a>bcbioSingleCell 0.1.4 (2018-04-30)</h1>
<ul>
<li>Updated <code><a href="../reference/sampleData.html">sampleData()</a></code> S4 methods to match update in bcbioBase. Now supports <code>clean</code> argument, which returns non-blacklisted factor columns only. See <code><a href="http://www.rdocumentation.org/packages/bcbioBase/topics/metadataBlacklist">bcbioBase::metadataBlacklist</a></code> for the blacklist.</li>
<li>Improved axis scale appearance on dimensionality reduction plots using <code><a href="http://www.rdocumentation.org/packages/scales/topics/pretty_breaks">scales::pretty_breaks()</a></code> internally.</li>
<li>Added <code>grid</code> argument to plots, where applicable.</li>
<li>Renamed example dataset from <code>bcb_small</code> to <code>indrops_small</code>.</li>
<li>Removed unnecessary method support for <code>interestingGroups()</code> and <code>metadata()</code>. These extend from <code>SummarizedExperiment</code> correctly now.</li>
<li>Fixed x-axis label centering for <code>plotCellCounts()</code> and <code>plotReadsPerCell()</code>.</li>
<li>Simplified seurat method support for <code>SingleCellExperiment</code>-like methods, where applicable. This includes <code>rowData</code>, <code>gene2symbol()</code>, and <code>interestingGroups()</code>.</li>
<li>Improved dark mode color support for <code><a href="../reference/plotDot.html">plotDot()</a></code>, <code><a href="../reference/plotFeature.html">plotFeatureTSNE()</a></code>, <code>plotMarker()</code>.</li>
<li>Updated sample metadata example in README.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-1-3-2018-04-25" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-3-2018-04-25" class="anchor"></a>bcbioSingleCell 0.1.3 (2018-04-25)</h1>
<div id="minor-changes-13" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-13" class="anchor"></a>Minor changes</h2>
<ul>
<li>Improved summary statistics output during <code>filterCells()</code> call.</li>
<li>Miscellaneous documentation improvements, most notably to <code><a href="../reference/bcbioSingleCell.html">bcbioSingleCell()</a></code> constructor function.</li>
<li>
<code><a href="../reference/plotViolin.html">plotViolin()</a></code> now uses a color border by default.</li>
<li>Improved <code>cell2sample</code> mapping internally for <code><a href="../reference/readCellRanger.html">readCellRanger()</a></code>.</li>
<li>Improved <a href="https://bioconductor.org">Bioconductor</a> 3.7 installation instructions.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-2-2018-04-24" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-2-2018-04-24" class="anchor"></a>bcbioSingleCell 0.1.2 (2018-04-24)</h1>
<div id="major-changes-7" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-7" class="anchor"></a>Major changes</h2>
<ul>
<li>Now using <code><a href="../reference/bcbioSingleCell.html">bcbioSingleCell()</a></code> instead of <code><a href="../reference/deprecated.html">loadSingleCell()</a></code> as the main constructor function to create a <code>bcbioSingleCell</code> object. <code><a href="../reference/deprecated.html">loadSingleCell()</a></code> is deprecated and still works, but will warn the user.</li>
<li>Renamed <code><a href="../reference/deprecated.html">loadCellRanger()</a></code> to <code><a href="../reference/readCellRanger.html">readCellRanger()</a></code> for better name consistency.</li>
<li>Quality control function color palettes now default to [ggplot2][] colors instead of using <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> palettes. This is defined using <code>scale_color_hue()</code> instead of <code>scale_color_viridis()</code> for example. The <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palette is still used by default for marker expression plots.</li>
<li>Use “<code>aggregate</code>” instead of “<code>sampleNameAggregate</code>” to define aggregate/grouped samples in metadata.</li>
</ul>
</div>
<div id="minor-changes-14" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-14" class="anchor"></a>Minor changes</h2>
<ul>
<li>Reexporting relevant [ggplot2][] and <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palettes.</li>
<li>Renamed references to “inDrops” from “inDrop”, where applicable.</li>
<li>Consistently using <code>sym()</code> in place of <code>.data</code> internally for tidy code.</li>
<li>Updated <code>seurat</code> blacklist for <code><a href="../reference/sampleData.html">sampleData()</a></code> generic.</li>
<li>
<code>plotCellTypesPerCluster()</code> and <code><a href="../reference/plotMarker.html">plotMarkerTSNE()</a></code> now use an automatic color palette by default, which enables for dynamic color palette support when <code>dark = TRUE</code>. Internally this is handled with the <code>theme_midnight()</code> and <code>theme_paperwhite()</code> [ggplot2][] themes.</li>
<li>Updated installation instructions to support <a href="https://bioconductor.org">Bioconductor</a> 3.7.</li>
</ul>
</div>
<div id="internal-changes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#internal-changes-1" class="anchor"></a>Internal changes</h2>
<ul>
<li>
<code><a href="../reference/metrics.html">metrics()</a></code> method support now defaults to <code>matrix</code> and works similarly for <code>dgCMatrix</code> sparse matrices. This is used in place of <code><a href="../reference/deprecated.html">calculateMetrics()</a></code> to generate the per cell quality control metrics.</li>
<li>Reworked and improved <code><a href="../reference/aggregateReplicates.html">aggregateReplicates()</a></code> internal code.</li>
<li>Fixed facet wrapping when <code>aggregate</code> is defined in metadata for quality control plots.</li>
<li>Improved internal code for <code>sanitizeMarkers()</code> to use map the gene annotations from <code>rowRanges()</code> better.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-1-2018-04-16" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-1-2018-04-16" class="anchor"></a>bcbioSingleCell 0.1.1 (2018-04-16)</h1>
<div id="major-changes-8" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-8" class="anchor"></a>Major changes</h2>
<ul>
<li>Added support for calculating <code>barcodeRanks()</code> and <code>barcodeRanksPerSample()</code>.</li>
<li>Now exporting <code>plotMarker()</code> in addition to <code><a href="../reference/defunct.html">plotMarkers()</a></code>.</li>
<li>Primary counts matrix slotted into <code>assay()</code> is named <code>counts</code> instead of <code>raw</code>, for better consistency with <code>SingleCellExperiment</code> class. The <code>counts()</code> generic requires that the primary assay slot is named <code>counts</code> to work correctly. Nothing else here has changed, just the name.</li>
<li>
<code><a href="../reference/deprecated.html">loadCellRanger()</a></code> now returns a <code>SingleCellExperiment</code> object instead of a <code>bcbioSingleCell</code> object.</li>
<li>Added support for <code>transgeneNames</code> and <code>spikeNames</code> when loading up a dataset.</li>
<li>Datasets from a poorly annotated genome can now be loaded up using <code>organism = NULL</code> during the <code><a href="../reference/deprecated.html">loadSingleCell()</a></code> call.</li>
<li>Methods now dispatch on <code>SingleCellExperiment</code> rather than <code>bcbioSingleCell</code> where applicable, providing support for <code>SingleCellExperiment</code> objects created elsewhere.</li>
</ul>
</div>
<div id="minor-changes-15" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-15" class="anchor"></a>Minor changes</h2>
<ul>
<li>Added support for labeling barcode ranks, such as elbow or inflection point on UMI counts per cell plots.</li>
<li>Added support for plotting metrics using an empirical distribution function (ECDF) plot.</li>
<li>Use <code>theme_midnight()</code> and <code>theme_paperwhite()</code> internally for dimensionality reduction plots.</li>
<li>TSNE and PCA plots now use an aspect ratio of 1 by default.</li>
<li>Improved imports from Matrix, S4Vectors, and methods.</li>
<li>Switched back to using base <code>stop()</code>, <code>warning()</code>, and <code>message()</code>.</li>
<li>
<code><a href="../reference/deprecated.html">inflectionPoint()</a></code> has been made defunct, in favor of using <code>barcodeRanks()</code>.</li>
<li>Moved internal constructors into S4 methods, where applicable.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-1-0-2018-04-04" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-1-0-2018-04-04" class="anchor"></a>bcbioSingleCell 0.1.0 (2018-04-04)</h1>
<div id="major-changes-9" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-9" class="anchor"></a>Major changes</h2>
<ul>
<li>
<code>bcbioSingleCell</code> S4 class now extends <code>SingleCellExperiment</code> instead of <code>SummarizedExperiment</code>. This requires definition of <code>rowRanges()</code> inside the object instead of <code>rowData()</code>. Similar functionality was added to the bcbioRNASeq package. Upgrade support will be provided using <code><a href="../reference/updateObject.html">updateObject()</a></code>in a future release.</li>
<li>Added a differential expression utility function named <code>diffExp()</code>, which uses <a href="https://github.com/statOmics/zingeR">zingeR</a>/[edgeR][] internally to calculate gene expression changes across cell groups.</li>
</ul>
</div>
<div id="minor-changes-16" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-16" class="anchor"></a>Minor changes</h2>
<ul>
<li>Added <code><a href="../reference/deprecated.html">plotCumulativeUMIsPerCell()</a></code> utility. This may be removed in a future update in favor of adding this plot into <code>plotUMIsPerCell()</code> using an ECDF plot.</li>
<li>Now using <code><a href="../reference/readCellTypeMarkers.html">readCellTypeMarkers()</a></code> to load marker <code>data.frames</code>, instead of <code><a href="../reference/deprecated.html">readCellTypeMarkersFile()</a></code>. This matches the conventions used in the <a href="http://bioinformatics.sph.harvard.edu/bcbioBase">bcbioBase</a> package.</li>
</ul>
</div>
</div>
    <div id="bcbiosinglecell-0-0-32-2018-03-12" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-32-2018-03-12" class="anchor"></a>bcbioSingleCell 0.0.32 (2018-03-12)</h1>
<ul>
<li>Fixes for object subsetting and <code><a href="../reference/deprecated.html">loadSingleCell()</a></code> organism calls</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-31-2018-02-21" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-31-2018-02-21" class="anchor"></a>bcbioSingleCell 0.0.31 (2018-02-21)</h1>
<ul>
<li>Improved handling of old <a href="https://www.ensembl.org">Ensembl</a> release version for <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> output.</li>
<li>Switched <a href="http://rmarkdown.rstudio.com">R Markdown</a> templates to consistent snake case formatting for variables and file names.</li>
<li>
<code><a href="../reference/prepareSingleCellTemplate.html">prepareSingleCellTemplate()</a></code> now uses <code>_setup.R</code> instead of <code>setup.R</code>.</li>
<li>Updated unit tests to work with new assert checks.</li>
<li>Export <code><a href="../reference/mapCellsToSamples.html">mapCellsToSamples()</a></code> instead of <code>cell2sample()</code>. <code>cell2sample()</code> now simply acts as an accessor function, returning the internally stored cell2sample mappings rather than trying to calculate. <code><a href="../reference/mapCellsToSamples.html">mapCellsToSamples()</a></code> performs the actual mapping from cellular barcodes to sample identifiers.</li>
<li>Export <code>metricsPerSample()</code>
</li>
<li>Now using <code><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/bpmapply">BiocParallel::bpmapply()</a></code> to loop across the sparse matrix files per sample internally in the <code>.sparseCountsList()</code> function, which is shared between <code><a href="../reference/deprecated.html">loadSingleCell()</a></code> and <code><a href="../reference/deprecated.html">loadCellRanger()</a></code>.</li>
<li>Updated assert checks.</li>
<li>Updated <code><a href="../reference/prepareSingleCellTemplate.html">prepareSingleCellTemplate()</a></code> to explicitly state which files to include for each R Markdown template, rather than inheriting from <code><a href="http://www.rdocumentation.org/packages/bcbioBase/topics/prepareTemplate">bcbioBase::prepareTemplate()</a></code>.</li>
<li>
<code><a href="../reference/selectSamples.html">selectSamples()</a></code> now fails on a sample mismatch, rather than warning.</li>
<li>Improved internal code for subset method, adding a <code>drop = FALSE</code> argument to the cellular barcode matching call.</li>
<li>Made all file loads in working examples a single line.</li>
<li>Simplified <a href="http://rmarkdown.rstudio.com">R Markdown</a> directory paths.</li>
<li>Made all loads in unit tests a single line.</li>
<li>Renamed <code>programs</code> metadata slot to <code>programVersions</code>, to improve consistency with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package.</li>
<li>Updated internal sample metadata sanitization to ensure all columns are factors.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-30-2018-01-26" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-30-2018-01-26" class="anchor"></a>bcbioSingleCell 0.0.30 (2018-01-26)</h1>
<ul>
<li>Fixes for QC plot labeling of bcbioSingleCell object with different per sample filtering cutoffs applied.</li>
<li>Added new <code><a href="../reference/fetchData.html">fetchGeneData()</a></code> function, that wraps the functionality of <code><a href="http://www.rdocumentation.org/packages/Seurat/topics/FetchData">Seurat::FetchData()</a></code> for specific genes.</li>
<li>Improved internal dimensionality reduction plotting code. The main function has been renamed to <code>.plotDR()</code> internally.</li>
<li>Simplified the code for <code><a href="../reference/fetchData.html">fetchTSNEExpressionData()</a></code> to return a standard <code>data.frame</code> with the cellular barcodes as rows, instead of the previously grouped <code>tibble</code> method. Now this function returns aggregate gene marker calculations in the <code>mean</code>, <code>median</code>, and <code>sum</code> columns. Since this method has way fewer rows than the grouped <code>tibble</code>, the [ggplot2][] code for <code><a href="../reference/plotMarker.html">plotMarkerTSNE()</a></code> now runs faster.</li>
<li>Explicitly declare <code>viridis::</code> for color palettes, where applicable.</li>
<li>
<code>colorPoints</code> argument has been renamed to <code>expression</code> for <code><a href="../reference/plotMarker.html">plotMarkerTSNE()</a></code>.</li>
<li>Dynamic gene symbol to ensgene conversion has been removed from the plotting functions, for greater simplicity. Now the <code>genes</code> argument simply matches against the rownames in the counts matrix of the object.</li>
<li>Decreased the default <code>minCumPct</code> argument for <code>plotPCElbow()</code> from 0.9 to 0.8. This is more conservative and will return slightly fewer principal components for dimensionality reduction, by default.</li>
<li>
<code><a href="../reference/plotDimensionalReduction.html">plotTSNE()</a></code>, <code>plotPCA()</code>, and the other dimensionality reduction-related plotting functions now default to a smaller point size (0.5) and slight alpha transparency (0.8), to make super imposed points more obvious for large datasets with many cells.</li>
<li>Added a working example for <code>subsetPerSample()</code>.</li>
<li>Internally switched from <code>.onLoad()</code> to <code>.onAttach()</code> method for automatically loading required dependency packages.</li>
<li>
<code>plotPCA()</code> now uses <code>phase</code> instead of <code>Phase</code> plotting cell cycle regression as an interestingGroup (see Seurat clustering template). Previously some of the <a href="http://satijalab.org/seurat">Seurat</a> metadata columns were not consistently sanitized to lowerCamelCase (e.g. <code>Phase</code>, <code>res.0.8</code>, <code>orig.ident</code>).</li>
<li>Suppress package startup messages in <a href="http://rmarkdown.rstudio.com">R Markdown</a> templates.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-29-2018-01-24" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-29-2018-01-24" class="anchor"></a>bcbioSingleCell 0.0.29 (2018-01-24)</h1>
<ul>
<li>Switched to <a href="http://rlang.r-lib.org">rlang</a> methods for errors, messages, and warnings: <code>abort()</code>, <code>inform()</code>, and <code>warn()</code>.</li>
<li>Updated <code>filterCells()</code> function to enable per sample filtering cutoffs. This works by passing in a named numeric vector, where the names must match the internal <code>sampleID</code> metadata column (not <code>sampleName</code>).</li>
<li>Improved internal sanitization of metrics available with the <code><a href="../reference/metrics.html">metrics()</a></code> accessor. Now all count columns (e.g. <code>nUMI</code>) are consistently integers, and all character vector columns are consistently coerced to factors.</li>
<li>Seurat metadata available through the bcbioSingleCell generics are now consistently sanitized in lowerCamelCase. This applies to <code>orig.ident</code> and the <code>res.*</code> metadata columns.</li>
<li>Explicit integers are now consistently used in all of the function parameter arguments.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-28-2018-01-22" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-28-2018-01-22" class="anchor"></a>bcbioSingleCell 0.0.28 (2018-01-22)</h1>
<ul>
<li>Manually define functions used to read barcodes and matrices. This improves the functionality of the internal <code>.readSparseCounts()</code> function.</li>
<li>Improved sample directory matching for <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> output.</li>
<li>Fixed sample metadata subsetting based on cell2sample factor levels in the internal subset code.</li>
<li>Added tabbed histogram, violin, and barplots for appropriate quality control functions in the <a href="http://rmarkdown.rstudio.com">R Markdown</a> code.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-27-2018-01-18" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-27-2018-01-18" class="anchor"></a>bcbioSingleCell 0.0.27 (2018-01-18)</h1>
<ul>
<li>Migrated core dependency imports from <a href="http://steinbaugh.com/basejump">basejump</a> to <a href="http://bioinformatics.sph.harvard.edu/bcbioBase">bcbioBase</a>.</li>
<li>Improved <code>colnames</code> and <code>rownames</code> handling for internal <code>.readSparseCounts()</code> function.</li>
<li>Reworked <code><a href="../reference/deprecated.html">loadCellRanger()</a></code> function. <code>refDataDir</code> parameter has been renamed to <code>refdataDir</code>.</li>
<li>Added <code>organism</code> and <code>genomeBuild</code> options to <code><a href="../reference/deprecated.html">loadSingleCell()</a></code>, to override the metadata set in the bcbio run, if necessary.</li>
<li>Improved if statement data class checks, where applicable.</li>
<li>Renamed internal <code>.sparseCountsTx2Gene()</code> to <code>.transcriptToGeneLevelCounts()</code>.</li>
<li>Updated <code>geomean</code> bind method in <code><a href="../reference/fetchData.html">fetchTSNEExpressionData()</a></code>.</li>
<li>Changed <code>minNovelty</code> default from 0.8 to 0.75.</li>
<li>Improved seurat class support for <code><a href="../reference/plotDot.html">plotDot()</a></code>.</li>
<li>Improved parameter names for <code><a href="../reference/plotMarker.html">plotKnownMarkersDetected()</a></code>. Now uses <code>tsneColor</code>, <code>violinFill</code>, and <code>dotColor</code>. Also added <code>pointsAsNumbers</code> parameter.</li>
<li>Added <code>subtitle</code> parameter for <code><a href="../reference/plotMarker.html">plotMarkerTSNE()</a></code>.</li>
<li>Added <code>tsneColor</code>, <code>violinFill</code>, <code>dotColor</code>, and <code>dark</code> parameters for <code><a href="../reference/defunct.html">plotMarkers()</a></code>.</li>
<li>Improved looping method for <code><a href="../reference/plotMarker.html">plotTopMarkers()</a></code> so that it renders correctly in <a href="http://rmarkdown.rstudio.com">R Markdown</a> calls.</li>
<li>Added method support for <code><a href="../reference/plotViolin.html">plotViolin()</a></code>.</li>
<li>Improved internal <code>cell2sample</code> handling in subset method code.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-26-2017-12-18" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-26-2017-12-18" class="anchor"></a>bcbioSingleCell 0.0.26 (2017-12-18)</h1>
<ul>
<li>Renamed <code><a href="../reference/defunct.html">readMarkersFile()</a></code> to <code><a href="../reference/deprecated.html">readCellTypeMarkersFile()</a></code>.</li>
<li>Improved internal handling of multiplexed CellRanger samples. These are count matrices with cellular barcodes ending in <code>-2</code>, for example.</li>
<li>Updated <code><a href="../reference/aggregateReplicates.html">aggregateReplicates()</a></code> code to work with basejump generic, which uses <code>groupings</code> instead of <code>cells</code> as the grouping parameter.</li>
<li>Added method support for <code>detectOrganism()</code>.</li>
<li>Improved filtering parameter output in <code>filterCells()</code>.</li>
<li>Updated <code>gene2symbol()</code> method support for bcbioSingleCell and seurat objects.</li>
<li>Fixed working example in <code>knownMarkersDetected()</code>.</li>
<li>Reworked internal code for <code>plotCellTypesPerCluster()</code>.</li>
<li>Improved internal checks for facet wrapping and color palette parameter arguments, where applicable.</li>
<li>Offloaded base <code>plotQuantileHeatmap()</code> functionality into basejump, for use in <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package.</li>
<li>Added unit test data for `loadSingleCell().</li>
<li>Added additional unit tests to improve code coverage.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-25-2017-12-11" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-25-2017-12-11" class="anchor"></a>bcbioSingleCell 0.0.25 (2017-12-11)</h1>
<ul>
<li>Prepared a minimal working example dataset.</li>
<li>Added some initial unit tests.</li>
<li>Renamed <code><a href="../reference/defunct.html">plotKnownMarkers()</a></code> to <code><a href="../reference/plotMarker.html">plotKnownMarkersDetected()</a></code>.</li>
<li>Renamed <code><a href="../reference/defunct.html">readMarkers()</a></code> to <code><a href="../reference/defunct.html">readMarkersFile()</a></code>.</li>
<li>Added <code>gene2symbol()</code> method support.</li>
<li>Moved <code><a href="../reference/plotDot.html">plotDot()</a></code> generic to basejump package.</li>
<li>Added internal <code>.checkFormat()</code> function, which will check for <code>ensgene</code> or <code>symbol</code> input.</li>
<li>Added internal <code>.convertGenesToSymbols()</code> utility function, for mapping Ensembl gene identifiers to gene symbols.</li>
<li>Use explicit calls to Seurat functions interally, for clarity.</li>
<li>Simplified bcbioSingleCell object return in <code><a href="../reference/deprecated.html">loadSingleCell()</a></code> function.</li>
<li>Added seurat class method support for <code>annotable()</code> function.</li>
<li>Added <code>bcbio&lt;-</code> assignment method support for seurat class objects.</li>
<li>
<code><a href="../reference/deprecated.html">calculateMetrics()</a></code> function now uses <code>annotable = TRUE</code> as default, instead of using <code>missing()</code> method.</li>
<li>Added method support for <code>cell2sample()</code> for seurat class objects.</li>
<li>
<code>cellTypesPerCluster()</code> now uses <code>min</code> and <code>max</code> default arguments that don’t remove any rows.</li>
<li>Added method support for seurat class objects to <code>counts()</code> function. This defaults to returning the raw counts (<code>normalized = FALSE</code>), but can also return log-normalized (<code>normalized = TRUE</code>) and scaled (<code>normalized = "scaled"</code>) counts.</li>
<li>Added Ensembl gene identifier mapping support to <code><a href="../reference/fetchData.html">fetchTSNEExpressionData()</a></code>.</li>
<li>
<code>filterCells()</code> now simply works in a destructive manner. We’ve removed the <code>drop</code> parameter. The messages displayed to the user during this function call have been improved, and now include more statistics on the step where the majority of cells are filtered.</li>
<li>Initial commit of <code>gene2symbol()</code> method support for bcbioSingleCell and seurat class objects.</li>
<li>Improved <code>interestingGroups&lt;-</code> assignment method support for seurat class objects.</li>
<li>Color palettes now default to viridis instead of inferno palette, where applicable.</li>
<li>Added Ensembl gene identifier support for <code><a href="../reference/plotDot.html">plotDot()</a></code> function.</li>
<li>
<code><a href="../reference/plotFeature.html">plotFeatureTSNE()</a></code> now uses a plural <code>features</code> parameter instead of <code>feature</code>, which is consistent with the syntax used in the other functions.</li>
<li>Added Ensembl gene identifier support to <code><a href="../reference/plotMarker.html">plotMarkerTSNE()</a></code>. The <code>format</code> argument still defaults to “symbol”, for consistency with previous behavior. However, in the future we recommend that users pass in stable Ensembl gene identifiers here if possible, for better reproducibility.</li>
<li>
<code><a href="../reference/defunct.html">plotMarkers()</a></code> now supports Ensembl gene identifiers.</li>
<li>
<code>plotPCElbow()</code> now silently returns the sequence of principal components (PCs) that we recommend to use for dimensionality reduction.</li>
<li>We’re now using “glm” instead of “gam” for <code>geom_smooth()</code> plotting, where applicable. See the <code><a href="../reference/plotQC.html">plotQC()</a></code> function code.</li>
<li>Improved the defaults for <code>plotQuantileHeatmap()</code> to enable faster plotting. Now the dendrogram calculations are skipped by default, which take a long time for large datasets.</li>
<li>Removed the draft <code>plotStressGenes()</code> function for now. Will try to add this in a future update.</li>
<li>Fixed Markdown header handling for <code><a href="../reference/plotMarker.html">plotTopMarkers()</a></code> if <code>headerLevel = NULL</code>.</li>
<li>Simplified <code><a href="../reference/selectSamples.html">selectSamples()</a></code> code to rely upon output of our bracket-based subsetting method. See <code>subset.R</code> file for more details.</li>
<li>Improved metadata update in bracket-based subsetting method.</li>
<li>
<code>subsetPerSample()</code> function now defaults to saving in the working directory.</li>
<li>Updated <code>topBarcodes()</code> function to rank by <code>nUMI</code> instead of <code>nCount</code> column, so it works with data from either bcbioSingleCell or seurat objects.</li>
<li>Minor tweaks to seurat coercion from bcbioSingleCell object. Here we’ve improved the metadata slotting.</li>
<li>Initial commit of example data script in the <code>data-raw/</code> directory!</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-24-2017-11-27" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-24-2017-11-27" class="anchor"></a>bcbioSingleCell 0.0.24 (2017-11-27)</h1>
<ul>
<li>Raw cellular barcodes are now slotted in <code>object<a href='http://github.com/cellularBarcodes'>@cellularBarcodes</a></code> as a <code>data.frame</code> instead of a per sample <code>list</code>. This makes downstream subsetting operations on the barcodes simpler.</li>
<li>Bug fixes for <code>cell2sample</code> mapping.</li>
<li>Switched back to stable CRAN version of roxygen2 (6.0.1) for documentation.</li>
<li>Renamed <code><a href="../reference/defunct.html">pcCutoff()</a></code> to <code>plotPCElbow()</code>. The function now returns a PC sequence from 1 to the cutoff (e.g. 1:10) instead of just the final PC cutoff value. The R Markdown clustering template has been updated to reflect this change.</li>
<li>Renamed <code><a href="../reference/defunct.html">quantileHeatmap()</a></code> to <code>plotQuantileHeatmap()</code>, for consistency with other plotting functions.</li>
<li>Initial support for customized bracket based subsetting, which now acts upon the raw cellular barcode counts stashed in the <code>object<a href='http://github.com/bcbio'>@bcbio</a></code> slot.</li>
<li>Moved <code><a href="../reference/defunct.html">darkTheme()</a></code> to <a href="http://steinbaugh.com/basejump">basejump</a> package and reworked as <code>midnightTheme()</code>, with improved colors and axis appearance.</li>
<li>Added <code>pointsAsNumbers</code> parameter to <code><a href="../reference/plotDimensionalReduction.html">plotTSNE()</a></code> and <code>plotPCA()</code> functions, to match the functionality in <code><a href="../reference/plotMarker.html">plotMarkerTSNE()</a></code>.</li>
<li>Overhauled <code><a href="../reference/deprecated.html">loadCellRanger()</a></code> to support multiplexed <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> matrix output. <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> adds a numeric suffix to the end of multiplexed barcodes (e.g. <code>AAACCTGGTTTACTCT-1</code> denotes cellular barcode <code>AAACCTGGTTTACTCT</code> is assigned to sample <code>1</code>).</li>
<li>Improved <code>cell2sample</code> mapping in <code><a href="../reference/aggregateReplicates.html">aggregateReplicates()</a></code> function, which uses the <code>sampleNameAggregate</code> column in sample metadata to define the aggregate sample pairings. The <code>summarize()</code> step at line 101 is slow for datasets with many samples and should be changed in the future to speed things up.</li>
<li>Improved internal <code>cell2sample()</code> code to handle <code>NULL</code> stashed mappings better.</li>
<li>Updated TNSE plotting functions to use <code>midnightTheme()</code> instead of <code><a href="../reference/defunct.html">darkTheme()</a></code>.</li>
<li>Added user-defined point and label sizes for <code><a href="../reference/plotMarker.html">plotMarkerTSNE()</a></code>.</li>
<li>Fixed typo in <code>plotMitoRatio()</code> where <code>maxGenes</code> cutoff was plotted instead of <code>maxMitoRatio</code>.</li>
<li>Added <code>legend</code> parameter argument to <code><a href="../reference/plotQC.html">plotQC()</a></code> function. Also improved handling of <code>NULL</code> return for <code>plotReadsPerCell()</code>, which can happen with <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> output.</li>
<li>Updated facet wrapping in <code>plotZeroesVsDepth()</code> to match the behavior in the other plotting functions.</li>
<li>Initial methods support for custom bracket-based subsetting.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-23-2017-11-22" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-23-2017-11-22" class="anchor"></a>bcbioSingleCell 0.0.23 (2017-11-22)</h1>
<ul>
<li>Improved facet wrapping of aggregated samples (<code>sampleNameAggregate</code> present in sample metadata), but removing code support for wrapping by multiplexed FASTQ description.</li>
<li>Simplified handling of <code>bcbioSingleCell</code> objects with <code>filterCells()</code> applied. This information is stored in the <code>metadata()</code> slot as 3 variables: (1) <code>filterParams</code>, numeric vector of the parameters used to define the cell filtering cutoffs; (2) <code>filterCells</code>, character vector of the cellular barcode IDs that passed filtering; (3) <code>filterGenes</code>, character vector of the <a href="https://www.ensembl.org">Ensembl</a> gene identifiers that have passed filtering, as determined by the <code>minCellsPerGene</code> parameter.</li>
<li>For <code>filterCells()</code> return, we’re now defaulting to a destructive operation, where the columns (cells) and rows (genes) of the object are adjusted to match the cells and genes that have passed filtering. Currently this can be adjusted with the <code>drop</code> argument for testing, but should generally be left as <code>drop = TRUE</code>.</li>
<li>We’re now slotting a <code>cell2sample</code> named factor in the <code>metadata()</code> slot, which makes downstream quality control operations faster. This is generated on the fly for previously saved objects that don’t have a stashed <code>cell2sample</code>.</li>
<li>Initial commit of <code><a href="../reference/plotQC.html">plotQC()</a></code> utility function, which plots multiple quality control functions for easy visualization. This defaults to output as a cowplot grid (<code>return = "grid"</code>), but can alternatively be set to return <a href="http://rmarkdown.rstudio.com">R Markdown</a> code (<code>return = "markdown"</code>).</li>
<li>
<code><a href="../reference/aggregateReplicates.html">aggregateReplicates()</a></code> operation has been improved to properly slot raw cellular barcodes in <code>object<a href='http://github.com/bcbio'>@bcbio</a>$cellularBarcodes</code>. The <code>filterCells</code> vector is adjusted, and <code>sampleMetadata</code> factors should be properly releveled.</li>
<li>The <code>counts()</code> accessor simply returns the sparse matrix contained in the <code>assay()</code> slot. The <code>filterCells</code> argument has been removed.</li>
<li>Messages have been added to the <code>filterCells()</code> function, to help the user determine at which step the majority of cells are being filtered. We’re keeping a non-destructive option using <code>drop = FALSE</code> for the time being, but this will likely be removed for improved simplicity in a future update.</li>
<li>Updated the internal code for <code><a href="../reference/metrics.html">metrics()</a></code> to use a simpler join operation on the <code>colData</code>, <code>cell2sample</code> and <code>sampleMetadata</code>.</li>
<li>Updated facet wrap code in quality control plots to not facet multiplexed FASTQ descriptions and simply check for <code>sampleNameAggregate</code>.</li>
<li>Improved appearance of <code>plotReadsPerCell()</code> labels and legends. Additionally, <code>plotReadsPerCell()</code> more efficiently handles the stashed values in the <code>nCount</code> column of <code>colData</code>, for faster plotting that having to rely on manipulation of the raw <code>cellularBarcodes</code> list stashed in <code>object<a href='http://github.com/bcbio'>@bcbio</a>$cellularBarcodes</code>.</li>
<li>
<code>sampleMetadata()</code> return is now consistently sanitized for <code>bcbioSingleCell</code> and <code>seurat</code> objects.</li>
<li>Minor tweaks to quality control template and setup.R files.</li>
<li>Added <code><a href="../reference/plotFeature.html">plotFeatureTSNE()</a></code> utility function. This improves on <code><a href="http://www.rdocumentation.org/packages/Seurat/topics/FeaturePlot">Seurat::FeaturePlot()</a></code> and enables the user to overlay the cluster identifiers on top of the t-SNE plot. <code><a href="../reference/defunct.html">plotFeatures()</a></code> is now deprecated in favor of this function.</li>
<li>Improved internal handling of Seurat data in the <code>.fetchDimDataSeurat()</code> function. This now keeps the cell ID as the rowname.</li>
<li>Allow the user to define the color palette (<code>color</code>), as well as <code>pointSize</code> and <code>labelSize</code> for <code>plotPCA()</code> and <code><a href="../reference/plotDimensionalReduction.html">plotTSNE()</a></code>.</li>
<li>Factors are now correctly releveled in <code>cell2sample()</code> return.</li>
<li>Improved internal code for <code><a href="../reference/fetchData.html">fetchTSNEExpressionData()</a></code>.</li>
<li>Bug fix for <code><a href="../reference/metrics.html">metrics()</a></code> accessor not including the cell ID as rownames.</li>
<li>Clustering template fixes. Now uses <code><a href="../reference/plotFeature.html">plotFeatureTSNE()</a></code> to assess quality control metrics on t-SNE.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-22-2017-11-17" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-22-2017-11-17" class="anchor"></a>bcbioSingleCell 0.0.22 (2017-11-17)</h1>
<ul>
<li>Now internally stashing a <code>cell2sample</code> data.frame, which helps speed up operations on cellular barcode metrics calculations for quality control plots.</li>
<li>Improved support for optional <code>annotable</code>, <code>ensemblVersion</code>, <code>gtfFile</code>, and <code>sampleMetadataFile</code> arguments in <code><a href="../reference/deprecated.html">loadSingleCell()</a></code> function.</li>
<li>Simplified some of the messages shown during sample loading, in an attempt to make them clearer and more informative. Also improved messages shown to the user during a <code>filterCells()</code> function call.</li>
<li>The <code><a href="../reference/metrics.html">metrics()</a></code> function will now look for a stashed <code>cell2sample</code> <code>data.frame</code>, which speeds up operations for quality control plots.</li>
<li>Improved handling of sample metadata columns as factors. In particular, levels should be correctly updated using <code>droplevels</code> in a <code><a href="../reference/selectSamples.html">selectSamples()</a></code> call. The <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package has also been updated to work in a similar fashion, where all columns in the sample metadata data.frame are now defined as factors.</li>
<li>Simplified <code>bcbioSingleCell</code> to <code>seurat</code> object coercion to stash all of the bcbio metadata, and simply return the basic <code>seurat</code> object, rather than trying to also perform normalization and scaling. These steps have instead been added back to the Seurat R Markdown clustering template.</li>
<li>Updated cell cycle and cell type markers from our master copy on <a href="https://www.google.com/sheets">Google Sheets</a>.</li>
<li>Added a troubleshooting section to the GitHub README, with a note on maximum DLLs.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-21-2017-11-08" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-21-2017-11-08" class="anchor"></a>bcbioSingleCell 0.0.21 (2017-11-08)</h1>
<ul>
<li>Updated package imports to match <a href="https://bioconductor.org">Bioconductor</a> 3.6.</li>
<li>Initial support for <code>plotCellTypesPerCluster()</code>.</li>
<li>Initial support for <code>plotMitoVsCoding()</code>. I broke this code out from <code>plotMitoRatio()</code>. We could opt to keep this in <code>plotMitoRatio</code> with a <code>geom = "scatterplot"</code> argument.</li>
<li>Initial commit of <code>.applyFilterCutoffs()</code> internal function, used to subset the object to contain only cells and genes that have passed quailty control filtering.</li>
<li>Improved <a href="http://satijalab.org/seurat">Seurat</a> <code>FindAllMarkers()</code> sanitization.</li>
<li>Made the quality control plots more modular. Now they support multiple geoms, including <code>boxplot</code>, <code>histogram</code>, <code>ridgeline</code>, and <code>violin</code> (default). Median labels are applied with the internal <code>.medianLabels()</code> function.</li>
<li>Updated error message for <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> directory structure.</li>
<li>Cellular barcode columns are no longer split with an underscore. Instead, they are kept as a single ACGT string. We’re now generating <code>cellID</code> to <code>sampleID</code> matching with a different method. In the future, we’ll stash a cell2sample data.frame inside the object, that makes this operation faster than the current <code>mclapply()</code> code.</li>
<li>Updated annotable support in <code><a href="../reference/deprecated.html">loadSingleCell()</a></code> and <code><a href="../reference/deprecated.html">loadCellRanger()</a></code>.</li>
<li>Added support for handling both gene- and transcript-level counts. The updated release of the bcbio single-cell pipeline now outputs at gene level.</li>
<li>Initial quality control plot support for seurat objects.</li>
<li>Added support for return of only filtered cells and genes with the <code>counts(filterCells = TRUE)</code> function.</li>
<li>Added assignment support for <code>interestingGroups&lt;-</code>.</li>
<li>Initial support for sample metadata generation from seurat object.</li>
<li>Improved internal code for <code><a href="../reference/selectSamples.html">selectSamples()</a></code>.</li>
<li>Updated <code>topMarkers()</code> to match Seurat v2.1 update.</li>
<li>Improved <code>bcbioSingleCell</code> to <code>seurat</code> coercion method with <code>setAs()</code>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-20-2017-10-24" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-20-2017-10-24" class="anchor"></a>bcbioSingleCell 0.0.20 (2017-10-24)</h1>
<ul>
<li>Upgraded to <a href="http://steinbaugh.com/basejump">basejump</a> 0.1.0 and <a href="http://satijalab.org/seurat">Seurat</a> 2.1.0 dependencies.</li>
<li>Improved documentation of NAMESPACE imports per function.</li>
<li>Switched to base grep functions where applicable (<code>grepl()</code>, <code>gsub()</code>).</li>
<li>Use GTF in package documentation rather than GFF. Applies primarily to the <code><a href="../reference/deprecated.html">loadSingleCell()</a></code> import function.</li>
<li>Restrict class support in S4 methods to <code>bcbioSingleCell</code>. Legacy <code>bcbioSCDataSet</code> class can be upgraded to <code>bcbioSingleCell</code> class using <code><a href="../reference/coerce.html">as(bcb, "bcbioSingleCell")</a></code> coercion.</li>
<li>Use filtered cell output for <code><a href="../reference/metrics.html">metrics()</a></code> and quality control functions by default.</li>
<li>Updated the quality control R Markdown to include <code>filterCells = FALSE</code> where applicable.</li>
<li>Draft support for aggregated technical replicates in quality control functions using <code>sampleNameAggregate</code> column in sample metadata. This doesn’t change the actual counts values. It only applies to visualization in the quality control plots currently.</li>
<li>Miscellaneous <a href="http://rmarkdown.rstudio.com">R Markdown</a> template updates. Primarily improvements to the setup chunk object loading workflow.</li>
<li>Removed lintr checks from testthat. This is breaking <code><a href="http://www.rdocumentation.org/packages/devtools/topics/test">devtools::test()</a></code>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-19-2017-10-12" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-19-2017-10-12" class="anchor"></a>bcbioSingleCell 0.0.19 (2017-10-12)</h1>
<ul>
<li>Renamed main object class from <code>bcbioSCDataSet</code> to <code>bcbioSingleCell</code>.</li>
<li>Cell filtering with <code>filterCells()</code> will now slot a named logical vector into <code>metadata(object)[["filteredCells"]]</code>, which will be used to dynamically subset the slotted internal <code>SummarizedExperiment</code> data. Now that we’re using this approach, we can return a modified <code>bcbioSingleCell</code> object rather than defining a separate <code>bcbioSCFiltered</code> class.</li>
<li>Renamed <code><a href="../reference/defunct.html">loadSingleCellRun()</a></code> to <code><a href="../reference/deprecated.html">loadSingleCell()</a></code>, to match <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package.</li>
<li>Now allowing implicit integers in our function code.</li>
<li>Added support for plotting technical replicates. This is handled by <code>sampleNameAggregate</code> in the sample metadata.</li>
<li>Now using ridgeline plots in place of histograms where applicable.</li>
<li>Travis CI checks take too long when loading SummarizedExperiment. Hopefully this will be fixed in the 3.6 release later this month.</li>
<li>New internal dark theme (<code><a href="../reference/defunct.html">darkTheme()</a></code>), based on the Seurat theme.</li>
<li>Initial commit of <code><a href="../reference/plotDot.html">plotDot()</a></code> function, based on <code><a href="http://www.rdocumentation.org/packages/Seurat/topics/DotPlot">Seurat::DotPlot()</a></code>.</li>
<li>Added new t-SNE plots that allow for consistent cluster labeling.</li>
<li>Providing legacy support for <code>bcbioSCDataSet</code> and <code>bcbioSCFiltered</code>, which will be deprecated in a future release.</li>
<li>Offloaded some internal code to basejump, for improved consistency with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package: <code>internal-projectDir.R</code>, <code>internal-readSampleMetadataFile.R</code>, <code>internal-sampleDirs.R</code>. We may want to provide this code as a shared bcbio core package (e.g. <a href="http://bioinformatics.sph.harvard.edu/bcbioBase">bcbioBase</a>) in the future.</li>
<li>Added internal utility to check for valid marker genes (<code>.validMarkers()</code>).</li>
<li>Improved <a href="https://www.ensembl.org">Ensembl</a> release version support (<code>ensemblVersion</code>).</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-18-2017-09-17" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-18-2017-09-17" class="anchor"></a>bcbioSingleCell 0.0.18 (2017-09-17)</h1>
<ul>
<li>Renamed <code><a href="../reference/defunct.html">plotClusters()</a></code> to <code><a href="../reference/defunct.html">plotMarkers()</a></code>. Added soft deprecation.</li>
<li>Added <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color support in t-SNE plots and heatmaps.</li>
<li>Converted <code><a href="../reference/defunct.html">loadSingleCellRun()</a></code> and <code><a href="../reference/deprecated.html">loadCellRanger()</a></code> from S4 generics back to standard functions.</li>
<li>Added t-SNE utility functions: <code><a href="../reference/fetchData.html">fetchTSNEData()</a></code>, <code><a href="../reference/fetchData.html">fetchTSNEExpressionData()</a></code>, and <code><a href="../reference/defunct.html">plotTSNEExpressionData()</a></code>. This enable plotting of geometric mean values of desired marker genes.</li>
<li>Updated NEWS to Markdown, with hyperlinks.</li>
<li>Offloaded generics that would otherwise conflict with bcbioRNASeq to the basejump package.</li>
<li>Improved <a href="https://cran.r-project.org/web/packages/roxygen2/vignettes/roxygen2.html">roxygen</a> documentation. Moved as much documentation as possible to the methods files.</li>
<li>Updated <code>cellCycleMarkers</code> and <code>cellTypeMarkers</code> data. Now supports Drosophila.</li>
<li>Sample IDs are now sanitized using <code>make.names()</code> instead of <code>camel()</code>. This avoids undesirable coercion of some IDs (e.g. <code>group1_1</code> into <code>group11</code>).</li>
<li>Added recommended package syntax guidelines.</li>
<li>lintr checks now allow implicit integers (e.g. <code>1</code> instead of <code>1L</code>).</li>
<li>Added <a href="http://satijalab.org/seurat">Seurat</a> as dependency in <code>DESCRIPTION</code> file. The package now attaches <a href="http://satijalab.org/seurat">Seurat</a> automatically.</li>
<li>Package no longer imports mononcle or suggests scater, scde, or scone. We’re planning on adding these back in a future update, but build checks on <a href="https://travis-ci.org">Travis CI</a> otherwise take too long.</li>
<li>Added new <code><a href="../reference/defunct.html">quantileHeatmap()</a></code> function.</li>
<li>Improved Markdown header support across functions, where applicable.</li>
<li>Improved <code>bcbioSCFiltered</code> to <code>seurat</code> coercion to slot relevant bcbio metadata.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-17-2017-09-03" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-17-2017-09-03" class="anchor"></a>bcbioSingleCell 0.0.17 (2017-09-03)</h1>
<ul>
<li>Renamed package from <code>bcbioSinglecell</code> to <code>bcbioSingleCell</code>.</li>
<li>Added <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palette support to quality control plots.</li>
<li>Added cell-cycle marker genes for human and mouse.</li>
<li>Slotted <code>organism</code> in <code>bcbioSCDataSet</code> metadata, in addition to <code>genomeBuild</code>.</li>
<li>Updated <code>bcbioSCFiltered</code> to <code>seurat</code> coercion method to also run <code>FindVariableGenes()</code> and <code>ScaleData()</code> by default.</li>
<li>Added cell-cycle regression into <a href="http://satijalab.org/seurat">Seurat</a> clustering [RMarkdown][] template.</li>
<li>Improved <a href="https://github.com/hadley/pkgdown">pkgdown</a> settings and website appearance.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-16-2017-08-25" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-16-2017-08-25" class="anchor"></a>bcbioSingleCell 0.0.16 (2017-08-25)</h1>
<ul>
<li>Support for CRAN release of <a href="http://satijalab.org/seurat">Seurat</a>.</li>
<li>Improved documentation of package NAMESPACE in <code>bcbioSinglecell-package.R</code> file.</li>
<li>Offloaded <code>download()</code> functionality to <a href="http://steinbaugh.com/basejump">basejump</a> package, to avoid collisions with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package. Function has been renamed to <code>externalFile()</code>.</li>
<li>Removed function deprecations to simplify the NAMESPACE.</li>
<li>Improved <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> sample matching. With these changes the internal <code>.detectPipeline()</code> function is no longer needed.</li>
<li>Improved sample directory matching in internal <code>.sampleDirs()</code> function.</li>
<li>Updated paths to <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> MatrixMarket files in internal <code>.readSparseCounts()</code> function.</li>
<li>Updated use of <code>packageSE()</code> to <code>prepareSE()</code>, matching the corresponding <a href="http://steinbaugh.com/basejump">basejump</a> function change.</li>
<li>Renamed internal use of <code>filter()</code> to <code>tidy_filter()</code>, to avoid future NAMESPACE collisions with <a href="http://bioconductor.org/packages/release/bioc/html/ensembldb.html">ensembldb</a> package.</li>
<li>Renamed <code>bcbioSCSubset</code> class to <code>bcbioSCFiltered</code> class.</li>
<li>Fixed memory issue in <code>plotZeroesVsDepth()</code> for datasets with high cell counts.</li>
<li>Improved sample matching for <code><a href="../reference/selectSamples.html">selectSamples()</a></code>. Will attempt to migrate this to bracket-based subsetting in a future update.</li>
<li>Restricted sample selection with <code><a href="../reference/selectSamples.html">selectSamples()</a></code> to only work on <code>bcbioSCFiltered</code> class for the time being. We can add bracket-based subsetting or S4 method support in <code><a href="../reference/selectSamples.html">selectSamples()</a></code> to properly work on <code>bcbioSCDataSet</code> class in a future update.</li>
<li>Updated [RMarkdown][] template settings and simplified the setup chunks.</li>
<li>Initial support for <code>bcbioSCFiltered</code> class coercion to <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a> <code>CellDataSet</code> class.</li>
<li>Suggest <a href="http://bioconductor.org/packages/release/bioc/html/scater.html">scater</a> and <a href="https://bioconductor.org/packages/release/bioc/html/scone.html">scone</a> packages for additional quality control and visualization.</li>
<li>Require <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a> &gt;= 2.5.0.</li>
<li>Added initial support for <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palette in quality control plots.</li>
<li>Suggest [rmarkdown][] and <a href="http://bioconductor.org/packages/release/bioc/html/scde.html">scde</a> packages.</li>
<li>Initial commit of <code>subsetPerSample()</code> function.</li>
<li>Miscellaneous [RMarkdown][] template improvements.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-15-2017-08-11" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-15-2017-08-11" class="anchor"></a>bcbioSingleCell 0.0.15 (2017-08-11)</h1>
<ul>
<li>Renamed functions in <code>lowerCamelCase</code> from <code>snake_case</code>.</li>
<li>Draft support for <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a>, <a href="http://bioconductor.org/packages/release/bioc/html/scater.html">scater</a>, and <a href="https://bioconductor.org/packages/release/bioc/html/scone.html">scone</a>.</li>
<li>Package now depends on <a href="https://www.bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a>.</li>
<li>Renamed <code>loadRun()</code> to <code><a href="../reference/defunct.html">loadSingleCellRun()</a></code> for improved compatibility with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package. This helps avoid NAMESPACE collisions between packages.</li>
<li>Improved support for custom GTF files.</li>
<li>Split out <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> import into a separate utility function named <code><a href="../reference/deprecated.html">loadCellRanger()</a></code>.</li>
<li>Added plotZerosVsDepth() by <a href='http://github.com/roryk'>@roryk</a>.</li>
<li>Renamed <code>filteringCriteria</code> to <code>filterParams</code> in <code><a href='http://github.com/metadata'>@metadata</a></code> slot.</li>
<li>Miscellaneous documentation fixes.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-14-2017-07-26" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-14-2017-07-26" class="anchor"></a>bcbioSingleCell 0.0.14 (2017-07-26)</h1>
<ul>
<li>Migrated plotting functions to S4 methods.</li>
<li>Improved sample metadata handling.</li>
<li>Updated sample name and cellular barcode sanization.</li>
<li>Initial commit of <a href="http://satijalab.org/seurat">Seurat</a> utility functions.</li>
<li>Initial commit of YAML parameters for [RMarkdown][] templates.</li>
<li>Draft support for <a href="http://satijalab.org/seurat">Seurat</a> v2 pre-release.</li>
<li>Initial support for <a href="https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/surecell-wta-ddseq.html">SureCell</a> UMIs.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-13-2017-07-10" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-13-2017-07-10" class="anchor"></a>bcbioSingleCell 0.0.13 (2017-07-10)</h1>
<ul>
<li>Integrated [bcbio][] and <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> workflows into <code>load_run()</code>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-12-2017-06-28" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-12-2017-06-28" class="anchor"></a>bcbioSingleCell 0.0.12 (2017-06-28)</h1>
<ul>
<li>Initial support for S4 object creation using <a href="https://www.bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-11-2017-06-24" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-11-2017-06-24" class="anchor"></a>bcbioSingleCell 0.0.11 (2017-06-24)</h1>
<ul>
<li>Added <a href="https://github.com/hadley/testthat">testthat</a>, <a href="https://github.com/jimhester/lintr">lintr</a>, and <a href="https://github.com/jimhester/covr">covr</a> support for code coverage.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-10-2017-06-15" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-10-2017-06-15" class="anchor"></a>bcbioSingleCell 0.0.10 (2017-06-15)</h1>
<ul>
<li>Updated [RMarkdown][] templates.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-9-2017-06-12" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-9-2017-06-12" class="anchor"></a>bcbioSingleCell 0.0.9 (2017-06-12)</h1>
<ul>
<li>Compatibility update for basejump S4 NAMESPACE changes.</li>
<li>Improved plots for quality control and filtering.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-8-2017-05-28" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-8-2017-05-28" class="anchor"></a>bcbioSingleCell 0.0.8 (2017-05-28)</h1>
<ul>
<li>Renamed mitochondrial plot functions.</li>
<li>Changed presentation of mitochondrial abundance as ratio instead of percentage.</li>
<li>Simplified NAMESPACE by offloading dependencies to basejump.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-7-2017-05-13" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-7-2017-05-13" class="anchor"></a>bcbioSingleCell 0.0.7 (2017-05-13)</h1>
<ul>
<li>Initial support for loading of 10x Genomics <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> output.</li>
<li>Add detection of droplet method based on the metadata file.</li>
<li>Draft support for import of <a href="https://1cell-bio.com">inDrop</a> i5 index barcode counts.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-6-2017-05-10" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-6-2017-05-10" class="anchor"></a>bcbioSingleCell 0.0.6 (2017-05-10)</h1>
<ul>
<li>Modified <code>load_run()</code> function for improved consistency with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package.</li>
<li>Initial incorporation of scater package into workflow.</li>
<li>Initial commit of RMarkdown template skeletons.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-5-2017-05-08" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-5-2017-05-08" class="anchor"></a>bcbioSingleCell 0.0.5 (2017-05-08)</h1>
<ul>
<li>Improvements to run object based on code in <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package.</li>
<li>Better handling in plots for datasets with many samples.</li>
<li>Labeled quality control cutoffs on the plots.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-4-2017-04-26" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-4-2017-04-26" class="anchor"></a>bcbioSingleCell 0.0.4 (2017-04-26)</h1>
<ul>
<li>Prepared package for <a href="http://www.tidyverse.org">tidyverse</a> dependency updates.</li>
<li>Updated HPC server detection.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-3-2017-04-20" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-3-2017-04-20" class="anchor"></a>bcbioSingleCell 0.0.3 (2017-04-20)</h1>
<ul>
<li>NAMESPACE consolidation and function cleanup.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-2-2017-04-12" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-2-2017-04-12" class="anchor"></a>bcbioSingleCell 0.0.2 (2017-04-12)</h1>
<ul>
<li>Added quality control plotting functions.</li>
<li>Converted all functions to standard evaluation.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-1-2017-03-01" class="section level1">
<h1 class="page-header">
<a href="#bcbiosinglecell-0-0-1-2017-03-01" class="anchor"></a>bcbioSingleCell 0.0.1 (2017-03-01)</h1>
<ul>
<li>Initial draft release.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bcbiosinglecell-0-1-17-2018-07-09">0.1.17</a></li>
        <li><a href="#bcbiosinglecell-0-1-16-2018-06-20">0.1.16</a></li>
        <li><a href="#bcbiosinglecell-0-1-15-2018-06-13">0.1.15</a></li>
        <li><a href="#bcbiosinglecell-0-1-14-2018-06-05">0.1.14</a></li>
        <li><a href="#bcbiosinglecell-0-1-13-2018-05-29">0.1.13</a></li>
        <li><a href="#bcbiosinglecell-0-1-12-2018-05-25">0.1.12</a></li>
        <li><a href="#bcbiosinglecell-0-1-11-2018-05-23">0.1.11</a></li>
        <li><a href="#bcbiosinglecell-0-1-10-2018-05-19">0.1.10</a></li>
        <li><a href="#bcbiosinglecell-0-1-9-2018-05-18">0.1.9</a></li>
        <li><a href="#bcbiosinglecell-0-1-8-2018-05-16">0.1.8</a></li>
        <li><a href="#bcbiosinglecell-0-1-7-2018-05-15">0.1.7</a></li>
        <li><a href="#bcbiosinglecell-0-1-6-2018-05-09">0.1.6</a></li>
        <li><a href="#bcbiosinglecell-0-1-5-2018-05-04">0.1.5</a></li>
        <li><a href="#bcbiosinglecell-0-1-4-2018-04-30">0.1.4</a></li>
        <li><a href="#bcbiosinglecell-0-1-3-2018-04-25">0.1.3</a></li>
        <li><a href="#bcbiosinglecell-0-1-2-2018-04-24">0.1.2</a></li>
        <li><a href="#bcbiosinglecell-0-1-1-2018-04-16">0.1.1</a></li>
        <li><a href="#bcbiosinglecell-0-1-0-2018-04-04">0.1.0</a></li>
        <li><a href="#bcbiosinglecell-0-0-32-2018-03-12">0.0.32</a></li>
        <li><a href="#bcbiosinglecell-0-0-31-2018-02-21">0.0.31</a></li>
        <li><a href="#bcbiosinglecell-0-0-30-2018-01-26">0.0.30</a></li>
        <li><a href="#bcbiosinglecell-0-0-29-2018-01-24">0.0.29</a></li>
        <li><a href="#bcbiosinglecell-0-0-28-2018-01-22">0.0.28</a></li>
        <li><a href="#bcbiosinglecell-0-0-27-2018-01-18">0.0.27</a></li>
        <li><a href="#bcbiosinglecell-0-0-26-2017-12-18">0.0.26</a></li>
        <li><a href="#bcbiosinglecell-0-0-25-2017-12-11">0.0.25</a></li>
        <li><a href="#bcbiosinglecell-0-0-24-2017-11-27">0.0.24</a></li>
        <li><a href="#bcbiosinglecell-0-0-23-2017-11-22">0.0.23</a></li>
        <li><a href="#bcbiosinglecell-0-0-22-2017-11-17">0.0.22</a></li>
        <li><a href="#bcbiosinglecell-0-0-21-2017-11-08">0.0.21</a></li>
        <li><a href="#bcbiosinglecell-0-0-20-2017-10-24">0.0.20</a></li>
        <li><a href="#bcbiosinglecell-0-0-19-2017-10-12">0.0.19</a></li>
        <li><a href="#bcbiosinglecell-0-0-18-2017-09-17">0.0.18</a></li>
        <li><a href="#bcbiosinglecell-0-0-17-2017-09-03">0.0.17</a></li>
        <li><a href="#bcbiosinglecell-0-0-16-2017-08-25">0.0.16</a></li>
        <li><a href="#bcbiosinglecell-0-0-15-2017-08-11">0.0.15</a></li>
        <li><a href="#bcbiosinglecell-0-0-14-2017-07-26">0.0.14</a></li>
        <li><a href="#bcbiosinglecell-0-0-13-2017-07-10">0.0.13</a></li>
        <li><a href="#bcbiosinglecell-0-0-12-2017-06-28">0.0.12</a></li>
        <li><a href="#bcbiosinglecell-0-0-11-2017-06-24">0.0.11</a></li>
        <li><a href="#bcbiosinglecell-0-0-10-2017-06-15">0.0.10</a></li>
        <li><a href="#bcbiosinglecell-0-0-9-2017-06-12">0.0.9</a></li>
        <li><a href="#bcbiosinglecell-0-0-8-2017-05-28">0.0.8</a></li>
        <li><a href="#bcbiosinglecell-0-0-7-2017-05-13">0.0.7</a></li>
        <li><a href="#bcbiosinglecell-0-0-6-2017-05-10">0.0.6</a></li>
        <li><a href="#bcbiosinglecell-0-0-5-2017-05-08">0.0.5</a></li>
        <li><a href="#bcbiosinglecell-0-0-4-2017-04-26">0.0.4</a></li>
        <li><a href="#bcbiosinglecell-0-0-3-2017-04-20">0.0.3</a></li>
        <li><a href="#bcbiosinglecell-0-0-2-2017-04-12">0.0.2</a></li>
        <li><a href="#bcbiosinglecell-0-0-1-2017-03-01">0.0.1</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

