<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single-Cell RNA-Seq Utilities • bcbioSingleCell</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Single-Cell RNA-Seq Utilities">
<meta property="og:description" content="Toolkit for bcbio single-cell RNA-seq analysis.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    <div id="bcbiosinglecell" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bcbiosinglecell" class="anchor"></a>bcbioSingleCell</h1></div>

<p><a href="https://www.r-project.org">R</a> package for <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> single-cell RNA-seq analysis.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="bioconductor-method" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor-method" class="anchor"></a><a href="https://bioconductor.org">Bioconductor</a> method</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## try http:// if https:// URLs are not supported
<span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"devtools"</span>)
<span class="kw">biocLite</span>(<span class="st">"remotes"</span>)
<span class="kw">biocLite</span>(<span class="st">"GenomeInfoDbData"</span>)
<span class="kw">biocLite</span>(
    <span class="st">"hbc/bcbioSingleCell"</span>,
    <span class="dt">dependencies =</span> <span class="kw">c</span>(<span class="st">"Depends"</span>, <span class="st">"Imports"</span>, <span class="st">"Suggests"</span>)
)</code></pre></div>
</div>
</div>
<div id="load-bcbio-run" class="section level2">
<h2 class="hasAnchor">
<a href="#load-bcbio-run" class="anchor"></a>Load <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> run</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(bcbioSingleCell)
bcb &lt;-<span class="st"> </span><span class="kw"><a href="reference/bcbioSingleCell.html">bcbioSingleCell</a></span>(
    <span class="dt">uploadDir =</span> <span class="st">"bcbio_indrop/final"</span>,
    <span class="dt">interestingGroups =</span> <span class="kw">c</span>(<span class="st">"genotype"</span>, <span class="st">"treatment"</span>),
    <span class="dt">sampleMetadataFile =</span> <span class="st">"sample_metadata.csv"</span>,
    <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>,
    <span class="dt">ensemblRelease =</span> 90L
)
<span class="co"># Back up all data inside bcbioSingleCell object</span>
flat &lt;-<span class="st"> </span><span class="kw">flatFiles</span>(bcb)
<span class="kw">saveData</span>(bcb, flat, <span class="dt">dir =</span> <span class="st">"data"</span>)</code></pre></div>
<p>This will return a <code>bcbioSingleCell</code> object, which is an extension of the <a href="https://bioconductor.org">Bioconductor</a> <a href="https://doi.org/doi:10.18129/B9.bioc.SingleCellExperiment">SingleCellExperiment</a> container class.</p>
<p>Parameters:</p>
<ul>
<li>
<code>uploadDir</code>: Path to the <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> final upload directory.</li>
<li>
<code>interestingGroups</code>: Character vector of the column names of interest in the sample metadata, which is stored in the <code><a href="reference/sampleData.html">sampleData()</a></code> accessor slot of the <code>bcbioSingleCell</code> object. These values should be formatted in camelCase, and can be reassigned in the object after creation (e.g. <code>interestingGroups(bcb) &lt;- c("batch", "age")</code>). They are used for data visualization in the quality control utility functions.</li>
<li>
<code>organism</code>: Organism name. Use the full latin name (e.g. “Homo sapiens”). If set <code>NULL</code>, no gene annotations will be downloaded and stashed into the <code>rowRanges</code> slot of the object.</li>
<li>
<code>genomeBuild</code>: <em>Optional.</em> The Ensembl release version to use.</li>
<li>
<code>gffFile</code>: <em>Optional.</em> If your transcriptome does not entirely match up to an Ensembl release, you can pass the GFF/GTF file of the transcriptome you used to load the annotations instead.</li>
</ul>
<p>Consult <code>help("bcbioSingleCell", "bcbioSingleCell")</code> for additional documentation.</p>
</div>
<div id="sample-metadata-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-metadata-examples" class="anchor"></a>Sample metadata examples</h2>
<div id="fastq-files-with-samples-multiplexed-by-index-barcode" class="section level3">
<h3 class="hasAnchor">
<a href="#fastq-files-with-samples-multiplexed-by-index-barcode" class="anchor"></a>FASTQ files with samples multiplexed by index barcode</h3>
<p>This is our current recommended method for analyzing an <a href="https://github.com/indrops/indrops">inDrops</a> dataset. The sample index barcodes are multiplexed per FASTQ set. For Illumina sequencing data, the raw binary base call (BCL) data must be converted into FASTQs (split into <code>R1</code>-<code>R4</code> files) using <a href="https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html">bcl2fastq</a>.</p>
<p>The <a href="https://github.com/indrops/indrops">inDrops</a> library version is automatically detected by <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a>, but ensure that the sample index sequences provided match the library version when attempting to create a <code>bcbioSingleCell</code> object. A current list of <a href="https://github.com/seqcloud/seqcloud/blob/master/workflows/bcbio/scrnaseq/harvard_indrop_v3/index_barcodes.csv">inDrops v3 index barcodes</a> is available from <a href="http://seq.cloud">seqcloud</a>.</p>
<p>Consult the <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> documentation for more information on how to configure an <a href="https://github.com/indrops/indrops">inDrops</a> run prior to loading into <a href="https://www.r-project.org">R</a> with the <code><a href="reference/bcbioSingleCell.html">bcbioSingleCell()</a></code> function.</p>
<table class="table">
<thead><tr class="header">
<th>description</th>
<th>index</th>
<th>sequence</th>
<th>sampleName</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>indrops1</td>
<td>17</td>
<td>GGAGGTAA</td>
<td>sample1</td>
</tr>
<tr class="even">
<td>indrops1</td>
<td>18</td>
<td>CATAACTG</td>
<td>sample2</td>
</tr>
<tr class="odd">
<td>indrops2</td>
<td>12</td>
<td>GCGTAAGA</td>
<td>sample3</td>
</tr>
<tr class="even">
<td>indrops2</td>
<td>13</td>
<td>CTATTAAG</td>
<td>sample4</td>
</tr>
<tr class="odd">
<td>indrops2</td>
<td>14</td>
<td>AAGGCTAT</td>
<td>sample5</td>
</tr>
<tr class="even">
<td>indrops2</td>
<td>15</td>
<td>GAGCCTTA</td>
<td>sample6</td>
</tr>
<tr class="odd">
<td>indrops2</td>
<td>16</td>
<td>TTATGCGA</td>
<td>sample7</td>
</tr>
</tbody>
</table>
</div>
<div id="fastq-files-demultiplexed-per-sample" class="section level3">
<h3 class="hasAnchor">
<a href="#fastq-files-demultiplexed-per-sample" class="anchor"></a>FASTQ files demultiplexed per sample</h3>
<p>This is our current method for handling <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">10X Genomics Cell Ranger</a> output (using <code><a href="reference/readCellRanger.html">readCellRanger()</a></code>) and <a href="https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/surecell-wta-ddseq.html">Illumina SureCell</a> sample data.</p>
<table class="table">
<thead><tr class="header">
<th>description</th>
<th>genotype</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>sample1</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>sample2</td>
<td>knockout</td>
</tr>
<tr class="odd">
<td>sample3</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>sample4</td>
<td>knockout</td>
</tr>
</tbody>
</table>
</div>
<div id="forbidden-metadata-fields" class="section level3">
<h3 class="hasAnchor">
<a href="#forbidden-metadata-fields" class="anchor"></a>Forbidden metadata fields</h3>
<p>The following fields are forbidden as user-supplied metadata, as they are used internally:</p>
<p><code>nCells</code></p>
</div>
</div>
<div id="troubleshooting" class="section level2">
<h2 class="hasAnchor">
<a href="#troubleshooting" class="anchor"></a>Troubleshooting</h2>
<div id="maximal-number-of-dlls-reached" class="section level3">
<h3 class="hasAnchor">
<a href="#maximal-number-of-dlls-reached" class="anchor"></a>Maximal number of DLLs reached</h3>
<pre><code>Error: package or namespace load failed for 'bcbioSingleCell' in dyn.load(file, DLLpath = DLLpath, ...):
  maximal number of DLLs reached...</code></pre>
<p>Depending on your operating system, you may encounter this error about hitting the DLL limit in <a href="https://www.r-project.org">R</a>. This issue is becoming more common as RNA-seq analysis packages grow increasingly complex. Luckily, we can configure <a href="https://www.r-project.org">R</a> to increase the DLL limit. Append this line to your <code>~/.Renviron</code> file:</p>
<pre><code>R_MAX_NUM_DLLS=150</code></pre>
<p>For more information on this issue, consult <code>help("dyn.load")</code> in the <a href="https://www.r-project.org">R</a> documentation. The number of loaded DLLs in an <a href="https://www.r-project.org">R</a> session can be obtained with <code>getLoadedDLLs()</code>.</p>
</div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/C8EMxl">shared library</a> on <a href="https://paperpile.com">Paperpile</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/hbc/bcbioSingleCell">https://​github.com/​hbc/​bcbioSingleCell</a>
</li>
<li>Report a bug at <br><a href="https://github.com/hbc/bcbioSingleCell/issues">https://​github.com/​hbc/​bcbioSingleCell/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://steinbaugh.com">Michael Steinbaugh</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2403-2221" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
<li>
<a href="https://github.com/roryk">Rory Kirchner</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4814-5885" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
<li>
<a href="https://github.com/marypiper">Mary Piper</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2699-3840" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
<li>
<a href="https://github.com/vbarrera">Victor Barrera</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-0590-4634" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
<li>
<a href="https://github.com/sjhosui">Shannan Ho Sui</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6191-4709" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
<li>
<a href="http://bioinformatics.sph.harvard.edu">Harvard Chan Bioinformatics Core</a> <br><small class="roles"> Copyright holder, funder </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/hbc/bcbioSingleCell"><img src="https://travis-ci.org/hbc/bcbioSingleCell.svg?branch=master" alt="Travis CI"></a></li>
<li><a href="https://codecov.io/gh/hbc/bcbioSingleCell"><img src="https://codecov.io/gh/hbc/bcbioSingleCell/branch/master/graph/badge.svg" alt="Codecov"></a></li>
<li><a href="https://www.repostatus.org/#active"><img src="https://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active - The project has reached a stable, usable state and is being actively developed."></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://steinbaugh.com">Michael Steinbaugh</a>, <a href="https://github.com/roryk">Rory Kirchner</a>, <a href="https://github.com/marypiper">Mary Piper</a>, <a href="https://github.com/vbarrera">Victor Barrera</a>, <a href="https://github.com/sjhosui">Shannan Ho Sui</a>, <a href="http://bioinformatics.sph.harvard.edu">Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
