<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single-Cell RNA-Seq Utilities • bcbioSingleCell</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">bcbioSingleCell</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    <div id="bcbiosinglecell" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bcbiosinglecell" class="anchor"></a>bcbioSingleCell</h1></div>
<p><a href="https://travis-ci.org/hbc/bcbioSingleCell"><img src="https://travis-ci.org/hbc/bcbioSingleCell.svg?branch=master" alt="Build Status"></a> <a href="http://www.repostatus.org/#wip"><img src="http://www.repostatus.org/badges/latest/wip.svg" alt="Project Status: WIP - Initial development is in progress, but there has not yet been a stable, usable release suitable for the public."></a> <a href="https://codecov.io/gh/hbc/bcbioSingleCell"><img src="https://codecov.io/gh/hbc/bcbioSingleCell/branch/master/graph/badge.svg" alt="codecov"></a></p>
<p>Import and analyze <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> single-cell RNA-seq data.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>This is an <a href="https://www.r-project.org">R</a> package.</p>
<div id="bioconductor-method" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor-method" class="anchor"></a><a href="https://bioconductor.org">Bioconductor</a> method</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(
    <span class="st">"hbc/bcbioSingleCell"</span>,
    <span class="dt">dependencies =</span> <span class="kw">c</span>(<span class="st">"Depends"</span>, <span class="st">"Imports"</span>, <span class="st">"Suggests"</span>)
)</code></pre></div>
</div>
</div>
<div id="sample-metadata-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-metadata-examples" class="anchor"></a>Sample metadata examples</h2>
<div id="fastq-files-with-samples-multiplexed-by-index-barcode" class="section level3">
<h3 class="hasAnchor">
<a href="#fastq-files-with-samples-multiplexed-by-index-barcode" class="anchor"></a>FASTQ files with samples multiplexed by index barcode</h3>
<p>This is our current method for handling inDrop samples.</p>
<table class="table">
<thead><tr class="header">
<th>fileName</th>
<th>description</th>
<th>index</th>
<th>sequence</th>
<th>sampleName</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>170201_R1.fastq.gz</td>
<td>run1</td>
<td>17</td>
<td>GGAGGTAA</td>
<td>sample1</td>
</tr>
<tr class="even">
<td>170201_R1.fastq.gz</td>
<td>run1</td>
<td>18</td>
<td>CATAACTG</td>
<td>sample2</td>
</tr>
<tr class="odd">
<td>170620_R1.fastq.gz</td>
<td>run2</td>
<td>12</td>
<td>GCGTAAGA</td>
<td>sample3</td>
</tr>
<tr class="even">
<td>170620_R1.fastq.gz</td>
<td>run2</td>
<td>13</td>
<td>CTATTAAG</td>
<td>sample4</td>
</tr>
<tr class="odd">
<td>170620_R1.fastq.gz</td>
<td>run2</td>
<td>14</td>
<td>AAGGCTAT</td>
<td>sample5</td>
</tr>
<tr class="even">
<td>170620_R1.fastq.gz</td>
<td>run2</td>
<td>15</td>
<td>GAGCCTTA</td>
<td>sample6</td>
</tr>
<tr class="odd">
<td>170620_R1.fastq.gz</td>
<td>run2</td>
<td>16</td>
<td>TTATGCGA</td>
<td>sample7</td>
</tr>
</tbody>
</table>
</div>
<div id="fastq-files-demultiplexed-per-sample" class="section level3">
<h3 class="hasAnchor">
<a href="#fastq-files-demultiplexed-per-sample" class="anchor"></a>FASTQ files demultiplexed per sample</h3>
<p>This is our current method for handling 10X and SureCell samples.</p>
<table class="table">
<thead><tr class="header">
<th>fileName</th>
<th>description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>sample1_170620_R1.fastq.gz</td>
<td>sample1</td>
</tr>
<tr class="even">
<td>sample2_170620_R1.fastq.gz</td>
<td>sample2</td>
</tr>
<tr class="odd">
<td>sample3_170620_R1.fastq.gz</td>
<td>sample3</td>
</tr>
<tr class="even">
<td>sample4_170620_R1.fastq.gz</td>
<td>sample4</td>
</tr>
<tr class="odd">
<td>sample5_170620_R1.fastq.gz</td>
<td>sample5</td>
</tr>
</tbody>
</table>
</div>
<div id="technical-replicates" class="section level3">
<h3 class="hasAnchor">
<a href="#technical-replicates" class="anchor"></a>Technical replicates</h3>
<p>Use <code>sampleNameAggregate</code> to assign groupings for technical replicates:</p>
<table class="table">
<thead><tr class="header">
<th>fileName</th>
<th>description</th>
<th>sampleNameAggregate</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>wildtype_L001_R1.fastq.gz</td>
<td>wildtype_L001</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>wildtype_L002_R1.fastq.gz</td>
<td>wildtype_L002</td>
<td>wildtype</td>
</tr>
<tr class="odd">
<td>wildtype_L003_R1.fastq.gz</td>
<td>wildtype_L003</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>wildtype_L004_R1.fastq.gz</td>
<td>wildtype_L004</td>
<td>wildtype</td>
</tr>
<tr class="odd">
<td>mutant_L001_R1.fastq.gz</td>
<td>mutant_L001</td>
<td>mutant</td>
</tr>
<tr class="even">
<td>mutant_L002_R1.fastq.gz</td>
<td>mutant_L002</td>
<td>mutant</td>
</tr>
<tr class="odd">
<td>mutant_L003_R1.fastq.gz</td>
<td>mutant_L003</td>
<td>mutant</td>
</tr>
<tr class="even">
<td>mutant_L004_R1.fastq.gz</td>
<td>mutant_L004</td>
<td>mutant</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="troubleshooting" class="section level2">
<h2 class="hasAnchor">
<a href="#troubleshooting" class="anchor"></a>Troubleshooting</h2>
<div id="maximal-number-of-dlls-reached" class="section level3">
<h3 class="hasAnchor">
<a href="#maximal-number-of-dlls-reached" class="anchor"></a>Maximal number of DLLs reached</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Error<span class="op">:</span><span class="st"> </span>package or namespace load failed <span class="cf">for</span> <span class="st">'bcbioSingleCell'</span> <span class="cf">in</span> <span class="kw">dyn.load</span>(file, <span class="dt">DLLpath =</span> DLLpath, ...)<span class="op">:</span>
<span class="st">  </span>maximal number of DLLs reached...</code></pre></div>
<p>Depending on your operating system, you may encounter this error about hitting the DLL limit in <a href="https://www.r-project.org">R</a>. This issue is becoming more common as RNA-seq analysis packages grow increasingly complex. Luckily, we can configure <a href="https://www.r-project.org">R</a> to increase the DLL limit. Append this line to your <code>~/.Renviron</code> file:</p>
<pre><code>R_MAX_NUM_DLLS=150</code></pre>
<p>For more information on this issue, consult <code>help("dyn.load")</code> in the <a href="https://www.r-project.org">R</a> documentation. The number of loaded DLLs in an <a href="https://www.r-project.org">R</a> session can be obtained with <code>getLoadedDLLs()</code>.</p>
</div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/C8EMxl">shared library</a> on <a href="https://paperpile.com">Paperpile</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3" id="sidebar">
    <h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/hbc/bcbioSingleCell">https://​github.com/​hbc/​bcbioSingleCell</a>
</li>
<li>Report a bug at <br><a href="https://github.com/hbc/bcbioSingleCell/issues">https://​github.com/​hbc/​bcbioSingleCell/​issues</a>
</li>
</ul>
<h2>License</h2>
<p><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE.html">LICENSE</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="http://steinbaugh.com">Michael Steinbaugh</a> <br><small class="roles"> Author, maintainer </small> </li>
<li>
<a href="https://github.com/roryk">Rory Kirchner</a> <br><small class="roles"> Author </small> </li>
<li>
<a href="https://github.com/marypiper">Mary Piper</a> <br><small class="roles"> Author </small> </li>
<li>
<a href="https://github.com/sjhosui">Shannan Ho Sui</a> <br><small class="roles"> Author </small> </li>
<li>
<a href="http://bioinformatics.sph.harvard.edu">Harvard Chan Bioinformatics Core</a> <br><small class="roles"> Copyright holder, funder </small> </li>
</ul>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://steinbaugh.com">Michael Steinbaugh</a>, <a href="https://github.com/roryk">Rory Kirchner</a>, <a href="https://github.com/marypiper">Mary Piper</a>, <a href="https://github.com/sjhosui">Shannan Ho Sui</a>, <a href="http://bioinformatics.sph.harvard.edu">Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
