<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title><code>bcbioSingleCell</code> Object and Constructor — bcbioSingleCell • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="<code>bcbioSingleCell</code> Object and Constructor — bcbioSingleCell" />

<meta property="og:description" content="bcbioSingleCell is an S4 class that extends SingleCellExperiment, and is
designed to store a bcbio single-cell RNA-seq analysis. This class contains
read counts saved as a sparse matrix (dgCMatrix), sample metadata, and cell
quality control metrics." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1><code>bcbioSingleCell</code> Object and Constructor</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/AllClasses.R'><code>R/AllClasses.R</code></a></small>
    <div class="hidden name"><code>bcbioSingleCell.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p><code>bcbioSingleCell</code> is an S4 class that extends <code>SingleCellExperiment</code>, and is
designed to store a bcbio single-cell RNA-seq analysis. This class contains
read counts saved as a sparse matrix (<code>dgCMatrix</code>), sample metadata, and cell
quality control metrics.</p>
    
    </div>

    <pre class="usage"><span class='fu'>bcbioSingleCell</span>(<span class='no'>uploadDir</span>, <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>, <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>spikeNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gffFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uploadDir</th>
      <td><p><code>string</code>. Path to final upload directory. This path is set
when running "<code>bcbio_nextgen -w template</code>".</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code>string</code> or <code>NULL</code>. Organism name. Use the full Latin name
(e.g. "Homo sapiens"), since this will be input downstream to AnnotationHub
and ensembldb, unless <code>gffFile</code> is set. If left <code>NULL</code> (<em>not recommended</em>),
the function call will skip loading gene-level annotations into
<code>rowRanges()</code>. This can be useful for poorly annotation genomes or
experiments involving multiple genomes.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p><code>string</code> or <code>NULL</code>. Sample barcode metadata file.
Optional for runs with demultiplixed index barcodes (e.g. SureCell) and can
be left <code>NULL</code>, but otherwise required for runs with multipliexed FASTQs
containing multiple index barcodes (e.g. inDrop).</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code> or <code>NULL</code>. Character vector of
interesting groups. Must be formatted in camel case and intersect with
<code><a href='sampleData.html'>sampleData()</a></code> colnames.</p></td>
    </tr>
    <tr>
      <th>ensemblRelease</th>
      <td><p><code>scalar integer</code> or <code>NULL</code>. Ensembl release version. If
<code>NULL</code>, defaults to current release, and does not typically need to be
user-defined. Passed to AnnotationHub for <code>EnsDb</code> annotation matching,
unless <code>gffFile</code> is set.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code>string</code> or <code>NULL</code>. Ensembl genome build name (e.g.
"GRCh38"). This will be passed to AnnotationHub for <code>EnsDb</code> annotation
matching, unless <code>gffFile</code> is set.</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code> or <code>NULL</code>. Vector indicating which
<code>assay()</code> rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code> or <code>NULL</code>. Vector indicating which <code>assay()</code>
rows denote spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p><code>string</code> or <code>NULL</code>. <em>Advanced use only; not recommended.</em> By
default, we recommend leaving this <code>NULL</code> for genomes that are supported on
Ensembl. In this case, the row annotations (<code>rowRanges()</code>) will be obtained
automatically from Ensembl by passing the <code>organism</code>, <code>genomeBuild</code>, and
<code>ensemblRelease</code> arguments to AnnotationHub and ensembldb. For a genome
that is not supported on Ensembl and/or AnnotationHub, a GFF/GTF (General
Feature Format) file is required. Generally, we recommend using a GTF
(GFFv2) file here over a GFF3 file if possible, although all GFF formats
are supported. The function will internally generate transcript-to-gene
mappings and construct a <code>GRanges</code> object containing the genomic ranges
(<code>rowRanges()</code>).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments, to be stashed in the <code>metadata()</code> slot.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>bcbioSingleCell</code>.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><code>bcbioSingleCell</code> extended <code>SummarizedExperiment</code> prior to v0.1.0,
where we migrated to <code>SingleCellExperiment</code>.</p>
    
    <h2 class="hasAnchor" id="remote-data"><a class="anchor" href="#remote-data"></a>Remote Data</h2>

    
    <p>When working in RStudio, we recommend connecting to the bcbio-nextgen
run directory as a remote connection over
<a href='https://github.com/osxfuse/osxfuse/wiki/SSHFS'>sshfs</a>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code><a href='http://www.rdocumentation.org/packages/SingleCellExperiment/topics/SingleCellExperiment'>SingleCellExperiment::SingleCellExperiment()</a></code>.</p></li>
<li><p><code>.S4methods(class = "bcbioSingleCell")</code>.</p></li>
</ul></div>
    <div class='dont-index'><p>Other S4 Object: <code><a href='extract.html'>extract</a></code>, <code><a href='show.html'>show</a></code>,
  <code><a href='updateObject.html'>updateObject</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>uploadDir</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata/indrops"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioSingleCell"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioSingleCell</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Homo sapiens"</span>,
    <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='no'>uploadDir</span>, <span class='st'>"metadata.csv"</span>),
    <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='fl'>87L</span>
)</div><div class='output co'>#&gt; <span class='message'>2018-01-01_bcbio</span></div><div class='output co'>#&gt; <span class='message'>1 samples detected:</span>
#&gt; <span class='message'>multiplexed_AAAAAAAA</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml</span></div><div class='output co'>#&gt; <span class='message'>Data versions are missing</span></div><div class='output co'>#&gt; <span class='message'>1000 reads per cellular barcode cutoff detected</span></div><div class='output co'>#&gt; <span class='message'>UMI type: surecell</span></div><div class='output co'>#&gt; <span class='message'>Reading metadata.csv</span></div><div class='output co'>#&gt; <span class='message'>Multiplexed samples detected</span></div><div class='output co'>#&gt; <span class='message'>Reading counts as genes</span></div><div class='output co'>#&gt; <span class='message'>Reading counts in mtx format</span></div><div class='output co'>#&gt; <span class='message'>Reading cellular barcode distributions</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.12.0 (2018-04-30)</span></div><div class='output co'>#&gt; <span class='message'>Ensembl 87 EnsDb for Homo Sapiens</span></div><div class='output co'>#&gt; <span class='message'>id: "AH53211"</span>
#&gt; <span class='message'>organism: "Homo sapiens"</span>
#&gt; <span class='message'>build: "GRCh38"</span>
#&gt; <span class='message'>release: "87"</span>
#&gt; <span class='message'>format: "genes"</span></div><div class='output co'>#&gt; <span class='message'>Standardizing the metadata columns</span></div><div class='output co'>#&gt; <span class='message'>Defining broad class using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Sorting ranges by geneID</span></div><div class='output co'>#&gt; <span class='message'>Calculating cellular barcode metrics</span></div><div class='output co'>#&gt; <span class='message'>500 cells detected</span></div><div class='output co'>#&gt; <span class='message'>449 coding genes</span></div><div class='output co'>#&gt; <span class='message'>13 mitochondrial genes</span></div><div class='output co'>#&gt; <span class='message'>500 / 500 cellular barcodes passed pre-filtering (100%)</span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; bcbioSingleCell 0.2.0
#&gt; class: SingleCellExperiment 
#&gt; dim: 500 500 
#&gt; metadata(31): version pipeline ... utilsSessionInfo devtoolsSessionInfo
#&gt; assays(1): counts
#&gt; rownames(500): ENSG00000000457 ENSG00000002586 ... ENSG00000283060
#&gt;   ENSG00000283391
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(500): multiplexed_AAAAAAAA_AAACACTA_ATCCCACG
#&gt;   multiplexed_AAAAAAAA_AAACACTA_CTTAGGTA ...
#&gt;   multiplexed_AAAAAAAA_TGTTATCA_ACGCAGAG
#&gt;   multiplexed_AAAAAAAA_TGTTATCA_CAAGTCAT
#&gt; colData names(15): nCount sampleID ... aggregate revcomp
#&gt; reducedDimNames(0):
#&gt; spikeNames(0):
#&gt; ====================================================================================================
#&gt; Upload Dir: "/Users/mike/GitHub/hbc/bcbioSingleCell/inst/extdata/indrops"
#&gt; Upload Date: 2018-01-01
#&gt; R Load Date: 2018-08-02
#&gt; Level: "genes"
#&gt; Organism: "Homo sapiens"
#&gt; Interesting Groups: "sampleName"
#&gt; Metadata File: "/Users/mike/GitHub/hbc/bcbioSingleCell/inst/extdata/indrops/metadata.csv"
#&gt; AnnotationHub: "AH53211"
#&gt; Ensembl Release: "87"
#&gt; Genome Build: "GRCh38"
#&gt; Filtered: FALSE
#&gt; zinbwave: FALSE</div><div class='input'><span class='fu'>validObject</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] TRUE</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#remote-data">Remote Data</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Rory Kirchner

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

