<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fetch Data Functions — fetchGeneData • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Fetch Data Functions — fetchGeneData" />

<meta property="og:description" content="Fetch Data Functions" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fetch Data Functions</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/fetchData-methods.R'><code>R/fetchData-methods.R</code></a></small>
    <div class="hidden name"><code>fetchData.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Fetch Data Functions</p>
    
    </div>

    <pre class="usage"><span class='fu'>fetchGeneData</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>fetchPCAData</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>fetchTSNEData</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>fetchTSNEExpressionData</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>fetchUMAPData</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>fetchUMAPExpressionData</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>fetchGeneData</span>(<span class='no'>object</span>, <span class='no'>genes</span>,
  <span class='kw'>gene2symbol</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>fetchGeneData</span>(<span class='no'>object</span>, <span class='no'>genes</span>, <span class='kw'>gene2symbol</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>fetchPCAData</span>(<span class='no'>object</span>, <span class='kw'>minimal</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>fetchPCAData</span>(<span class='no'>object</span>, <span class='kw'>minimal</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>fetchTSNEData</span>(<span class='no'>object</span>, <span class='kw'>minimal</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>fetchTSNEData</span>(<span class='no'>object</span>, <span class='kw'>minimal</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>fetchTSNEExpressionData</span>(<span class='no'>object</span>, <span class='no'>genes</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>fetchTSNEExpressionData</span>(<span class='no'>object</span>, <span class='no'>genes</span>)

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>fetchUMAPData</span>(<span class='no'>object</span>, <span class='kw'>minimal</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>fetchUMAPData</span>(<span class='no'>object</span>, <span class='kw'>minimal</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>fetchUMAPExpressionData</span>(<span class='no'>object</span>, <span class='no'>genes</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>fetchUMAPExpressionData</span>(<span class='no'>object</span>, <span class='no'>genes</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p><code>character</code>. Gene identifiers. Must match the rownames of the
object.</p></td>
    </tr>
    <tr>
      <th>gene2symbol</th>
      <td><p><code>data.frame</code>. Gene-to-symbol mappings. Columns must
contain <code>geneID</code> and <code>geneName</code>.</p></td>
    </tr>
    <tr>
      <th>minimal</th>
      <td><p><code>boolean</code>. Return minimal data without metrics.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    
<ul>
<li><p><code>fetchGeneData()</code>: <code>matrix</code>.</p></li>
<li><p><code>fetchTSNEExpressionData()</code>: <code>data.frame</code> containing t-SNE coordinates,
sample metadata, and aggregate marker expression values (<code>mean</code>, <code>median</code>,
and <code>sum</code>).</p></li>
<li><p>Other functions: <code>data.frame</code>, containing metrics.</p></li>
</ul>

    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Some of these functions require "<code>ident</code>" to be defined in <code>colData()</code>.
For t-SNE, UMAP, and PCA, the reduced dimensions must be defined in the
<code>reducedDims()</code> slot.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Data Functions: <code><a href='aggregateReplicates.html'>aggregateReplicates</a></code>,
  <code><a href='cell2sample.html'>cell2sample</a></code>,
  <code><a href='metricsPerSample.html'>metricsPerSample</a></code>, <code><a href='sampleData.html'>sampleData</a></code>,
  <code><a href='selectSamples.html'>selectSamples</a></code>,
  <code><a href='subsetPerSample.html'>subsetPerSample</a></code>, <code><a href='topBarcodes.html'>topBarcodes</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># SingleCellExperiment ====</span>
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>indrops_small</span>
<span class='no'>genes</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='fu'>rownames</span>(<span class='no'>object</span>))
<span class='fu'>glimpse</span>(<span class='no'>genes</span>)</div><div class='output co'>#&gt;  chr [1:6] "ENSG00000000457" "ENSG00000002586" "ENSG00000004939" ...</div><div class='input'>
<span class='co'># Genes</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>fetchGeneData</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>genes</span>)
<span class='fu'>glimpse</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt;  num [1:500, 1:6] 1 1 4 3 3 2 5 1 3 3 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:500] "multiplexed_AAAAAAAA_AAACACTA_ATCCCACG" "multiplexed_AAAAAAAA_AAACACTA_CTTAGGTA" "multiplexed_AAAAAAAA_AAACACTA_CTTCGATT" "multiplexed_AAAAAAAA_AAACCGCC_GAGTTTCG" ...
#&gt;   ..$ : chr [1:6] "ENSG00000000457" "ENSG00000002586" "ENSG00000004939" "ENSG00000005022" ...</div><div class='input'>
<span class='co'># t-SNE</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>fetchTSNEData</span>(<span class='no'>object</span>)
<span class='fu'>glimpse</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Observations: 500
#&gt; Variables: 24
#&gt; $ nGene            &lt;int&gt; 289, 405, 423, 337, 350, 378, 442, 217, 332, 405, ...
#&gt; $ nUMI             &lt;int&gt; 1199, 2766, 4399, 1485, 1771, 2208, 4095, 1406, 15...
#&gt; $ nCount           &lt;int&gt; 25114, 55337, 71802, 34607, 34005, 44760, 76590, 3...
#&gt; $ sampleID         &lt;fct&gt; multiplexed_AAAAAAAA, multiplexed_AAAAAAAA, multip...
#&gt; $ nCoding          &lt;int&gt; 1105, 2385, 3864, 1296, 1574, 1599, 3630, 695, 117...
#&gt; $ nMito            &lt;int&gt; 24, 241, 234, 103, 83, 452, 234, 556, 246, 225, 28...
#&gt; $ log10GenesPerUMI &lt;dbl&gt; 0.7992992, 0.7575732, 0.7208579, 0.7969256, 0.7832...
#&gt; $ mitoRatio        &lt;dbl&gt; 0.02001668, 0.08712943, 0.05319391, 0.06936027, 0....
#&gt; $ sampleName       &lt;fct&gt; rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, r...
#&gt; $ fileName         &lt;fct&gt; multiplexed.fastq.gz, multiplexed.fastq.gz, multip...
#&gt; $ description      &lt;fct&gt; multiplexed-AAAAAAAA, multiplexed-AAAAAAAA, multip...
#&gt; $ index            &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,...
#&gt; $ sequence         &lt;fct&gt; TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, ...
#&gt; $ aggregate        &lt;fct&gt; sample, sample, sample, sample, sample, sample, sa...
#&gt; $ revcomp          &lt;fct&gt; AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, ...
#&gt; $ orig.ident       &lt;fct&gt; multiplexed, multiplexed, multiplexed, multiplexed...
#&gt; $ res.0.4          &lt;chr&gt; "0", "2", "0", "0", "0", "1", "0", "0", "0", "2", ...
#&gt; $ res.0.8          &lt;chr&gt; "3", "2", "1", "1", "3", "0", "1", "1", "3", "2", ...
#&gt; $ res.1.2          &lt;chr&gt; "4", "0", "1", "1", "4", "2", "1", "1", "4", "0", ...
#&gt; $ ident            &lt;fct&gt; 0, 2, 0, 0, 0, 1, 0, 0, 0, 2, 0, 1, 1, 1, 0, 1, 2,...
#&gt; $ tSNE_1           &lt;dbl&gt; -17.8005980, 14.3021396, 5.5353587, 5.5824008, -17...
#&gt; $ tSNE_2           &lt;dbl&gt; -8.7635904, -1.2800499, -15.9179439, -9.7737431, -...
#&gt; $ centerX          &lt;dbl&gt; -1.020748, 13.796989, -1.020748, -1.020748, -1.020...
#&gt; $ centerY          &lt;dbl&gt; -9.92682961, 0.06908376, -9.92682961, -9.92682961,...</div><div class='input'>
<span class='co'># PCA</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>fetchPCAData</span>(<span class='no'>object</span>)
<span class='fu'>glimpse</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Observations: 500
#&gt; Variables: 24
#&gt; $ nGene            &lt;int&gt; 289, 405, 423, 337, 350, 378, 442, 217, 332, 405, ...
#&gt; $ nUMI             &lt;int&gt; 1199, 2766, 4399, 1485, 1771, 2208, 4095, 1406, 15...
#&gt; $ nCount           &lt;int&gt; 25114, 55337, 71802, 34607, 34005, 44760, 76590, 3...
#&gt; $ sampleID         &lt;fct&gt; multiplexed_AAAAAAAA, multiplexed_AAAAAAAA, multip...
#&gt; $ nCoding          &lt;int&gt; 1105, 2385, 3864, 1296, 1574, 1599, 3630, 695, 117...
#&gt; $ nMito            &lt;int&gt; 24, 241, 234, 103, 83, 452, 234, 556, 246, 225, 28...
#&gt; $ log10GenesPerUMI &lt;dbl&gt; 0.7992992, 0.7575732, 0.7208579, 0.7969256, 0.7832...
#&gt; $ mitoRatio        &lt;dbl&gt; 0.02001668, 0.08712943, 0.05319391, 0.06936027, 0....
#&gt; $ sampleName       &lt;fct&gt; rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, r...
#&gt; $ fileName         &lt;fct&gt; multiplexed.fastq.gz, multiplexed.fastq.gz, multip...
#&gt; $ description      &lt;fct&gt; multiplexed-AAAAAAAA, multiplexed-AAAAAAAA, multip...
#&gt; $ index            &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,...
#&gt; $ sequence         &lt;fct&gt; TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, ...
#&gt; $ aggregate        &lt;fct&gt; sample, sample, sample, sample, sample, sample, sa...
#&gt; $ revcomp          &lt;fct&gt; AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, ...
#&gt; $ orig.ident       &lt;fct&gt; multiplexed, multiplexed, multiplexed, multiplexed...
#&gt; $ res.0.4          &lt;chr&gt; "0", "2", "0", "0", "0", "1", "0", "0", "0", "2", ...
#&gt; $ res.0.8          &lt;chr&gt; "3", "2", "1", "1", "3", "0", "1", "1", "3", "2", ...
#&gt; $ res.1.2          &lt;chr&gt; "4", "0", "1", "1", "4", "2", "1", "1", "4", "0", ...
#&gt; $ ident            &lt;fct&gt; 0, 2, 0, 0, 0, 1, 0, 0, 0, 2, 0, 1, 1, 1, 0, 1, 2,...
#&gt; $ PC1              &lt;dbl&gt; 1.0316848, 5.3377977, 2.2791735, 3.3898010, 1.1814...
#&gt; $ PC2              &lt;dbl&gt; 1.39906902, -2.94220122, 0.22684589, 1.44953021, 0...
#&gt; $ centerX          &lt;dbl&gt; 1.962767, 3.819283, 1.962767, 1.962767, 1.962767, ...
#&gt; $ centerY          &lt;dbl&gt; 1.84123959, -2.55660965, 1.84123959, 1.84123959, 1...</div><div class='input'>
<span class='co'># UMAP</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>fetchUMAPData</span>(<span class='no'>object</span>)
<span class='fu'>glimpse</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Observations: 500
#&gt; Variables: 24
#&gt; $ nGene            &lt;int&gt; 289, 405, 423, 337, 350, 378, 442, 217, 332, 405, ...
#&gt; $ nUMI             &lt;int&gt; 1199, 2766, 4399, 1485, 1771, 2208, 4095, 1406, 15...
#&gt; $ nCount           &lt;int&gt; 25114, 55337, 71802, 34607, 34005, 44760, 76590, 3...
#&gt; $ sampleID         &lt;fct&gt; multiplexed_AAAAAAAA, multiplexed_AAAAAAAA, multip...
#&gt; $ nCoding          &lt;int&gt; 1105, 2385, 3864, 1296, 1574, 1599, 3630, 695, 117...
#&gt; $ nMito            &lt;int&gt; 24, 241, 234, 103, 83, 452, 234, 556, 246, 225, 28...
#&gt; $ log10GenesPerUMI &lt;dbl&gt; 0.7992992, 0.7575732, 0.7208579, 0.7969256, 0.7832...
#&gt; $ mitoRatio        &lt;dbl&gt; 0.02001668, 0.08712943, 0.05319391, 0.06936027, 0....
#&gt; $ sampleName       &lt;fct&gt; rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, r...
#&gt; $ fileName         &lt;fct&gt; multiplexed.fastq.gz, multiplexed.fastq.gz, multip...
#&gt; $ description      &lt;fct&gt; multiplexed-AAAAAAAA, multiplexed-AAAAAAAA, multip...
#&gt; $ index            &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,...
#&gt; $ sequence         &lt;fct&gt; TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, ...
#&gt; $ aggregate        &lt;fct&gt; sample, sample, sample, sample, sample, sample, sa...
#&gt; $ revcomp          &lt;fct&gt; AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, ...
#&gt; $ orig.ident       &lt;fct&gt; multiplexed, multiplexed, multiplexed, multiplexed...
#&gt; $ res.0.4          &lt;chr&gt; "0", "2", "0", "0", "0", "1", "0", "0", "0", "2", ...
#&gt; $ res.0.8          &lt;chr&gt; "3", "2", "1", "1", "3", "0", "1", "1", "3", "2", ...
#&gt; $ res.1.2          &lt;chr&gt; "4", "0", "1", "1", "4", "2", "1", "1", "4", "0", ...
#&gt; $ ident            &lt;fct&gt; 0, 2, 0, 0, 0, 1, 0, 0, 0, 2, 0, 1, 1, 1, 0, 1, 2,...
#&gt; $ UMAP1            &lt;dbl&gt; 1.42426342, 4.41265806, 3.12128718, 1.65304247, 1....
#&gt; $ UMAP2            &lt;dbl&gt; 1.2632623, 1.4619932, -5.3615592, -3.3089443, 0.88...
#&gt; $ centerX          &lt;dbl&gt; 1.390433, 4.982767, 1.390433, 1.390433, 1.390433, ...
#&gt; $ centerY          &lt;dbl&gt; -3.395173, 1.284410, -3.395173, -3.395173, -3.3951...</div><div class='input'>
<span class='co'># t-SNE gene expression</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>fetchTSNEExpressionData</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>genes</span>)
<span class='fu'>glimpse</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Observations: 500
#&gt; Variables: 27
#&gt; $ nGene            &lt;int&gt; 289, 405, 423, 337, 350, 378, 442, 217, 332, 405, ...
#&gt; $ nUMI             &lt;int&gt; 1199, 2766, 4399, 1485, 1771, 2208, 4095, 1406, 15...
#&gt; $ nCount           &lt;int&gt; 25114, 55337, 71802, 34607, 34005, 44760, 76590, 3...
#&gt; $ sampleID         &lt;fct&gt; multiplexed_AAAAAAAA, multiplexed_AAAAAAAA, multip...
#&gt; $ nCoding          &lt;int&gt; 1105, 2385, 3864, 1296, 1574, 1599, 3630, 695, 117...
#&gt; $ nMito            &lt;int&gt; 24, 241, 234, 103, 83, 452, 234, 556, 246, 225, 28...
#&gt; $ log10GenesPerUMI &lt;dbl&gt; 0.7992992, 0.7575732, 0.7208579, 0.7969256, 0.7832...
#&gt; $ mitoRatio        &lt;dbl&gt; 0.02001668, 0.08712943, 0.05319391, 0.06936027, 0....
#&gt; $ sampleName       &lt;fct&gt; rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, r...
#&gt; $ fileName         &lt;fct&gt; multiplexed.fastq.gz, multiplexed.fastq.gz, multip...
#&gt; $ description      &lt;fct&gt; multiplexed-AAAAAAAA, multiplexed-AAAAAAAA, multip...
#&gt; $ index            &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,...
#&gt; $ sequence         &lt;fct&gt; TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, ...
#&gt; $ aggregate        &lt;fct&gt; sample, sample, sample, sample, sample, sample, sa...
#&gt; $ revcomp          &lt;fct&gt; AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, ...
#&gt; $ orig.ident       &lt;fct&gt; multiplexed, multiplexed, multiplexed, multiplexed...
#&gt; $ res.0.4          &lt;chr&gt; "0", "2", "0", "0", "0", "1", "0", "0", "0", "2", ...
#&gt; $ res.0.8          &lt;chr&gt; "3", "2", "1", "1", "3", "0", "1", "1", "3", "2", ...
#&gt; $ res.1.2          &lt;chr&gt; "4", "0", "1", "1", "4", "2", "1", "1", "4", "0", ...
#&gt; $ ident            &lt;fct&gt; 0, 2, 0, 0, 0, 1, 0, 0, 0, 2, 0, 1, 1, 1, 0, 1, 2,...
#&gt; $ tSNE_1           &lt;dbl&gt; -17.8005980, 14.3021396, 5.5353587, 5.5824008, -17...
#&gt; $ tSNE_2           &lt;dbl&gt; -8.7635904, -1.2800499, -15.9179439, -9.7737431, -...
#&gt; $ centerX          &lt;dbl&gt; -1.020748, 13.796989, -1.020748, -1.020748, -1.020...
#&gt; $ centerY          &lt;dbl&gt; -9.92682961, 0.06908376, -9.92682961, -9.92682961,...
#&gt; $ mean             &lt;dbl&gt; 2.8333333, 1.8333333, 5.0000000, 1.3333333, 2.8333...
#&gt; $ median           &lt;dbl&gt; 1.0, 1.0, 2.0, 1.0, 1.5, 2.0, 4.0, 1.5, 1.5, 3.0, ...
#&gt; $ sum              &lt;dbl&gt; 17, 11, 30, 8, 17, 14, 34, 12, 10, 26, 14, 3, 10, ...</div><div class='input'>
<span class='co'># UMAP gene expession</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>fetchUMAPExpressionData</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>genes</span>)
<span class='fu'>glimpse</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Observations: 500
#&gt; Variables: 27
#&gt; $ nGene            &lt;int&gt; 289, 405, 423, 337, 350, 378, 442, 217, 332, 405, ...
#&gt; $ nUMI             &lt;int&gt; 1199, 2766, 4399, 1485, 1771, 2208, 4095, 1406, 15...
#&gt; $ nCount           &lt;int&gt; 25114, 55337, 71802, 34607, 34005, 44760, 76590, 3...
#&gt; $ sampleID         &lt;fct&gt; multiplexed_AAAAAAAA, multiplexed_AAAAAAAA, multip...
#&gt; $ nCoding          &lt;int&gt; 1105, 2385, 3864, 1296, 1574, 1599, 3630, 695, 117...
#&gt; $ nMito            &lt;int&gt; 24, 241, 234, 103, 83, 452, 234, 556, 246, 225, 28...
#&gt; $ log10GenesPerUMI &lt;dbl&gt; 0.7992992, 0.7575732, 0.7208579, 0.7969256, 0.7832...
#&gt; $ mitoRatio        &lt;dbl&gt; 0.02001668, 0.08712943, 0.05319391, 0.06936027, 0....
#&gt; $ sampleName       &lt;fct&gt; rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, rep_1, r...
#&gt; $ fileName         &lt;fct&gt; multiplexed.fastq.gz, multiplexed.fastq.gz, multip...
#&gt; $ description      &lt;fct&gt; multiplexed-AAAAAAAA, multiplexed-AAAAAAAA, multip...
#&gt; $ index            &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,...
#&gt; $ sequence         &lt;fct&gt; TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, TTTTTTTT, ...
#&gt; $ aggregate        &lt;fct&gt; sample, sample, sample, sample, sample, sample, sa...
#&gt; $ revcomp          &lt;fct&gt; AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, AAAAAAAA, ...
#&gt; $ orig.ident       &lt;fct&gt; multiplexed, multiplexed, multiplexed, multiplexed...
#&gt; $ res.0.4          &lt;chr&gt; "0", "2", "0", "0", "0", "1", "0", "0", "0", "2", ...
#&gt; $ res.0.8          &lt;chr&gt; "3", "2", "1", "1", "3", "0", "1", "1", "3", "2", ...
#&gt; $ res.1.2          &lt;chr&gt; "4", "0", "1", "1", "4", "2", "1", "1", "4", "0", ...
#&gt; $ ident            &lt;fct&gt; 0, 2, 0, 0, 0, 1, 0, 0, 0, 2, 0, 1, 1, 1, 0, 1, 2,...
#&gt; $ UMAP1            &lt;dbl&gt; 1.42426342, 4.41265806, 3.12128718, 1.65304247, 1....
#&gt; $ UMAP2            &lt;dbl&gt; 1.2632623, 1.4619932, -5.3615592, -3.3089443, 0.88...
#&gt; $ centerX          &lt;dbl&gt; 1.390433, 4.982767, 1.390433, 1.390433, 1.390433, ...
#&gt; $ centerY          &lt;dbl&gt; -3.395173, 1.284410, -3.395173, -3.395173, -3.3951...
#&gt; $ mean             &lt;dbl&gt; 2.8333333, 1.8333333, 5.0000000, 1.3333333, 2.8333...
#&gt; $ median           &lt;dbl&gt; 1.0, 1.0, 2.0, 1.0, 1.5, 2.0, 4.0, 1.5, 1.5, 3.0, ...
#&gt; $ sum              &lt;dbl&gt; 17, 11, 30, 8, 17, 14, 34, 12, 10, 26, 14, 3, 10, ...</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Rory Kirchner

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

