<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Filter Cells — filterCells • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Filter Cells — filterCells" />

<meta property="og:description" content="Apply gene detection, novelty score, and mitochondrial abundance cutoffs to
cellular barcodes. By default we recommend applying the same filtering cutoff
to all samples. The filtering parameters now support per-sample cutoffs,
defined using a named numeric vector. When matching per sample, be sure to
use the sampleNames() return values (i.e. the sampleName column in
sampleData())." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Filter Cells</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/filterCells-methods.R'><code>R/filterCells-methods.R</code></a></small>
    <div class="hidden name"><code>filterCells.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Apply gene detection, novelty score, and mitochondrial abundance cutoffs to
cellular barcodes. By default we recommend applying the same filtering cutoff
to all samples. The filtering parameters now support per-sample cutoffs,
defined using a named <code>numeric</code> vector. When matching per sample, be sure to
use the <code>sampleNames()</code> return values (i.e. the <code>sampleName</code> column in
<code><a href='sampleData.html'>sampleData()</a></code>).</p>
    
    </div>

    <pre class="usage"><span class='fu'>filterCells</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>filterCells</span>(<span class='no'>object</span>, <span class='kw'>nCells</span> <span class='kw'>=</span> <span class='fl'>Inf</span>,
  <span class='kw'>minUMIs</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>maxUMIs</span> <span class='kw'>=</span> <span class='fl'>Inf</span>, <span class='kw'>minGenes</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>maxGenes</span> <span class='kw'>=</span> <span class='fl'>Inf</span>,
  <span class='kw'>minNovelty</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>maxMitoRatio</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='kw'>minCellsPerGene</span> <span class='kw'>=</span> <span class='fl'>1L</span>,
  <span class='kw'>zinbwave</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>nCells</th>
      <td><p><code>scalar integer</code>. Expected number of cells per sample.</p></td>
    </tr>
    <tr>
      <th>minUMIs</th>
      <td><p><code>scalar integer</code>. Minimum number of UMI disambiguated counts
per cell.</p></td>
    </tr>
    <tr>
      <th>maxUMIs</th>
      <td><p><code>scalar integer</code>. Maximum number of UMI disambiguated counts
per cell.</p></td>
    </tr>
    <tr>
      <th>minGenes</th>
      <td><p><code>scalar integer</code>. Minimum number of genes detected.</p></td>
    </tr>
    <tr>
      <th>maxGenes</th>
      <td><p><code>scalar integer</code>. Maximum number of genes detected.</p></td>
    </tr>
    <tr>
      <th>minNovelty</th>
      <td><p><code>scalar integer</code> (<code>0</code>-<code>1</code>). Minimum novelty score (log10
genes per UMI).</p></td>
    </tr>
    <tr>
      <th>maxMitoRatio</th>
      <td><p><code>scalar integer</code> (<code>0</code>-<code>1</code>). Maximum relative
mitochondrial abundance.</p></td>
    </tr>
    <tr>
      <th>minCellsPerGene</th>
      <td><p><code>scalar integer</code>. Include genes with non-zero
expression in at least this many cells.</p></td>
    </tr>
    <tr>
      <th>zinbwave</th>
      <td><p><code>boolean</code>. Run <code><a href='http://www.rdocumentation.org/packages/zinbwave/topics/zinbwave'>zinbwave::zinbwave()</a></code> to automatically apply
a ZINB regression model to calculate <code>normalizedValues</code> and <code>weights</code>
matrices to by used for differential expression with DESeq2 or edgeR.
Note that this calculation should only be performed on <strong>filtered data</strong>.
For large datasets this can take a long time and use a lot of memory, so
this calculation is disabled by default.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>bcbioSingleCell</code>, with filtering information slotted into
<code>metadata()</code> as <code>filterCells</code> and <code>filterParams</code>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='http://www.rdocumentation.org/packages/Seurat/topics/CreateSeuratObject'>Seurat::CreateSeuratObject()</a></code>.</p>
<p>Other Quality Control Functions: <code><a href='barcodeRanksPerSample.html'>barcodeRanksPerSample</a></code>,
  <code><a href='metrics.html'>metrics</a></code>, <code><a href='plotCellCounts.html'>plotCellCounts</a></code>,
  <code><a href='plotGenesPerCell.html'>plotGenesPerCell</a></code>,
  <code><a href='plotMitoRatio.html'>plotMitoRatio</a></code>,
  <code><a href='plotMitoVsCoding.html'>plotMitoVsCoding</a></code>,
  <code><a href='plotNovelty.html'>plotNovelty</a></code>, <code><a href='plotQC.html'>plotQC</a></code>,
  <code><a href='plotReadsPerCell.html'>plotReadsPerCell</a></code>,
  <code><a href='plotUMIsPerCell.html'>plotUMIsPerCell</a></code>,
  <code><a href='plotZerosVsDepth.html'>plotZerosVsDepth</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># SingleCellExperiment ====</span>
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>cellranger_small</span>
<span class='fu'><a href='show.html'>show</a></span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; class: SingleCellExperiment 
#&gt; dim: 500 500 
#&gt; metadata(20): version pipeline ... utilsSessionInfo devtoolsSessionInfo
#&gt; assays(1): counts
#&gt; rownames(500): ENSG00000002586 ENSG00000005022 ... ENSG00000277734
#&gt;   ENSG00000279483
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(500): pbmc_1_AAACCTGCAGGCGATA pbmc_1_AAACCTGCATGAAGTA ...
#&gt;   pbmc_1_TTTGGTTTCGCTAGCG pbmc_1_TTTGTCACAGGTCCAC
#&gt; colData names(15): nGene nUMI ... res.1.2 ident
#&gt; reducedDimNames(3): PCA TSNE UMAP
#&gt; spikeNames(0):</div><div class='input'>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>filterCells</span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; <span class='message'>Parameters:</span>
#&gt; <span class='message'>  - &gt;= 0 UMIs per cell</span>
#&gt; <span class='message'>  - &lt;= Inf UMIs per cell</span>
#&gt; <span class='message'>  - &gt;= 0 genes per cell</span>
#&gt; <span class='message'>  - &lt;= Inf genes per cell</span>
#&gt; <span class='message'>  - &gt;= 0 novelty score</span>
#&gt; <span class='message'>  - &lt;= 1 mitochondrial abundance</span>
#&gt; <span class='message'>  - == Inf cells per sample</span>
#&gt; <span class='message'>  - &gt;= 1 cells per gene</span>
#&gt; <span class='message'>================================================================================</span>
#&gt; <span class='message'>Cells:</span>
#&gt; <span class='message'>     500 cells | prefilter</span>
#&gt; <span class='message'>     500 cells | minUMIs &gt;= 0</span>
#&gt; <span class='message'>     500 cells | maxUMIs &lt;= Inf</span>
#&gt; <span class='message'>     500 cells | minGenes &gt;= 0</span>
#&gt; <span class='message'>     500 cells | maxGenes &lt;= Inf</span>
#&gt; <span class='message'>     500 cells | minNovelty &lt;= 0</span>
#&gt; <span class='message'>     500 cells | maxMitoRatio &lt;= 1</span>
#&gt; <span class='message'>     500 cells | nCells == Inf</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>pbmc </span>
#&gt; <span class='message'> 500 </span>
#&gt; <span class='message'>================================================================================</span>
#&gt; <span class='message'>Genes:</span>
#&gt; <span class='message'>     500 genes | prefilter</span>
#&gt; <span class='message'>     500 genes | minCellsPerGene &lt;= 1</span>
#&gt; <span class='message'>================================================================================</span>
#&gt; <span class='message'>Summary:</span>
#&gt; <span class='message'>  - 500 of 500 cells passed filtering (100%)</span>
#&gt; <span class='message'>  - 500 of 500 genes passed filtering (100%)</span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: SingleCellExperiment 
#&gt; dim: 500 500 
#&gt; metadata(24): version pipeline ... filterParams filterSummary
#&gt; assays(1): counts
#&gt; rownames(500): ENSG00000002586 ENSG00000005022 ... ENSG00000277734
#&gt;   ENSG00000279483
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(500): pbmc_1_AAACCTGCAGGCGATA pbmc_1_AAACCTGCATGAAGTA ...
#&gt;   pbmc_1_TTTGGTTTCGCTAGCG pbmc_1_TTTGTCACAGGTCCAC
#&gt; colData names(15): nGene nUMI ... res.1.2 ident
#&gt; reducedDimNames(3): PCA TSNE UMAP
#&gt; spikeNames(0):</div><div class='input'><span class='fu'>metadata</span>(<span class='no'>x</span>)$<span class='no'>filterParams</span></div><div class='output co'>#&gt; $nCells
#&gt; [1] Inf
#&gt; 
#&gt; $minUMIs
#&gt; [1] 0
#&gt; 
#&gt; $maxUMIs
#&gt; [1] Inf
#&gt; 
#&gt; $minGenes
#&gt; [1] 0
#&gt; 
#&gt; $maxGenes
#&gt; [1] Inf
#&gt; 
#&gt; $minNovelty
#&gt; [1] 0
#&gt; 
#&gt; $maxMitoRatio
#&gt; [1] 1
#&gt; 
#&gt; $minCellsPerGene
#&gt; [1] 1
#&gt; </div><div class='input'>
<span class='co'># Per sample cutoffs</span>
<span class='fu'>sampleNames</span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; pbmc_1 
#&gt; "pbmc" </div><div class='input'><span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>filterCells</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
    <span class='kw'>minUMIs</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='kw'>pbmc</span> <span class='kw'>=</span> <span class='fl'>100</span>)
)</div><div class='output co'>#&gt; <span class='message'>minUMIs: per sample mode</span>
#&gt; <span class='message'>pbmc </span>
#&gt; <span class='message'> 100 </span></div><div class='output co'>#&gt; <span class='message'>Parameters:</span>
#&gt; <span class='message'>  - &gt;= 100 UMIs per cell</span>
#&gt; <span class='message'>  - &lt;= Inf UMIs per cell</span>
#&gt; <span class='message'>  - &gt;= 0 genes per cell</span>
#&gt; <span class='message'>  - &lt;= Inf genes per cell</span>
#&gt; <span class='message'>  - &gt;= 0 novelty score</span>
#&gt; <span class='message'>  - &lt;= 1 mitochondrial abundance</span>
#&gt; <span class='message'>  - == Inf cells per sample</span>
#&gt; <span class='message'>  - &gt;= 1 cells per gene</span>
#&gt; <span class='message'>================================================================================</span>
#&gt; <span class='message'>Cells:</span>
#&gt; <span class='message'>     500 cells | prefilter</span>
#&gt; <span class='message'>     500 cells | minUMIs &gt;= 100</span>
#&gt; <span class='message'>     500 cells | maxUMIs &lt;= Inf</span>
#&gt; <span class='message'>     500 cells | minGenes &gt;= 0</span>
#&gt; <span class='message'>     500 cells | maxGenes &lt;= Inf</span>
#&gt; <span class='message'>     500 cells | minNovelty &lt;= 0</span>
#&gt; <span class='message'>     500 cells | maxMitoRatio &lt;= 1</span>
#&gt; <span class='message'>     500 cells | nCells == Inf</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>pbmc </span>
#&gt; <span class='message'> 500 </span>
#&gt; <span class='message'>================================================================================</span>
#&gt; <span class='message'>Genes:</span>
#&gt; <span class='message'>     500 genes | prefilter</span>
#&gt; <span class='message'>     500 genes | minCellsPerGene &lt;= 1</span>
#&gt; <span class='message'>================================================================================</span>
#&gt; <span class='message'>Summary:</span>
#&gt; <span class='message'>  - 500 of 500 cells passed filtering (100%)</span>
#&gt; <span class='message'>  - 500 of 500 genes passed filtering (100%)</span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: SingleCellExperiment 
#&gt; dim: 500 500 
#&gt; metadata(24): version pipeline ... filterParams filterSummary
#&gt; assays(1): counts
#&gt; rownames(500): ENSG00000002586 ENSG00000005022 ... ENSG00000277734
#&gt;   ENSG00000279483
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(500): pbmc_1_AAACCTGCAGGCGATA pbmc_1_AAACCTGCATGAAGTA ...
#&gt;   pbmc_1_TTTGGTTTCGCTAGCG pbmc_1_TTTGTCACAGGTCCAC
#&gt; colData names(15): nGene nUMI ... res.1.2 ident
#&gt; reducedDimNames(3): PCA TSNE UMAP
#&gt; spikeNames(0):</div><div class='input'><span class='fu'>metadata</span>(<span class='no'>x</span>)$<span class='no'>filterParams</span></div><div class='output co'>#&gt; $nCells
#&gt; [1] Inf
#&gt; 
#&gt; $minUMIs
#&gt; pbmc 
#&gt;  100 
#&gt; 
#&gt; $maxUMIs
#&gt; [1] Inf
#&gt; 
#&gt; $minGenes
#&gt; [1] 0
#&gt; 
#&gt; $maxGenes
#&gt; [1] Inf
#&gt; 
#&gt; $minNovelty
#&gt; [1] 0
#&gt; 
#&gt; $maxMitoRatio
#&gt; [1] 1
#&gt; 
#&gt; $minCellsPerGene
#&gt; [1] 1
#&gt; </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

