<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>General Arguments — general • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="General Arguments — general" />

<meta property="og:description" content="General Arguments" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>General Arguments</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/general.R'><code>R/general.R</code></a></small>
    <div class="hidden name"><code>general.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>General Arguments</p>
    
    </div>

    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>aspectRatio</th>
      <td><p><code>scalar integer</code>. Aspect ratio.</p></td>
    </tr>
    <tr>
      <th>color</th>
      <td><p><code>ggproto</code>/<code>ScaleDiscrete</code> or <code>NULL</code>. Desired ggplot2 color
scale. Must supply discrete values. When set to <code>NULL</code>, the default ggplot2
color palette will be used. If manual color definitions are desired, we
recommend using <code><a href='http://www.rdocumentation.org/packages/ggplot2/topics/scale_manual'>ggplot2::scale_color_manual()</a></code>.
To set the discrete color palette globally, use
<code>options(bcbio.discrete.color = scale_color_viridis_d())</code>.</p></td>
    </tr>
    <tr>
      <th>fill</th>
      <td><p><code>ggproto</code>/<code>ScaleDiscrete</code> or <code>NULL</code>. Desired ggplot2 fill scale.
Must supply discrete values. When set to <code>NULL</code>, the default ggplot2 color
palette will be used. If manual color definitions are desired, we recommend
using <code><a href='http://www.rdocumentation.org/packages/ggplot2/topics/scale_manual'>ggplot2::scale_fill_manual()</a></code>.
To set the discrete fill palette globally, use
<code>options(bcbio.discrete.fill = scale_fill_viridis_d())</code>.</p></td>
    </tr>
    <tr>
      <th>gene2symbol</th>
      <td><p><code>data.frame</code>. Gene-to-symbol mappings. Columns must
contain <code>geneID</code> and <code>geneName</code>.</p></td>
    </tr>
    <tr>
      <th>dark</th>
      <td><p><code>boolean</code>. Plot against a dark background using
<code><a href='http://www.rdocumentation.org/packages/basejump/topics/themes'>basejump::theme_midnight()</a></code>.</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p><code>string</code>. Output directory path.</p></td>
    </tr>
    <tr>
      <th>expression</th>
      <td><p><code>string</code>. Calculation to apply. Uses <code>match.arg()</code> and
defaults to the first argument in the <code>character</code> vector.</p></td>
    </tr>
    <tr>
      <th>gene2symbol</th>
      <td><p><code>boolean</code>. Automatically interconvert gene identifiers
and gene names in plots. Useful for marker plots.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p><code>character</code>. Gene identifiers. Must match the rownames of the
object.</p></td>
    </tr>
    <tr>
      <th>geom</th>
      <td><p><code>string</code>. Plot type. Uses <code>match.arg()</code> and defaults to the first
argument in the <code>character</code> vector.</p></td>
    </tr>
    <tr>
      <th>grid</th>
      <td><p><code>boolean</code>. Show major grid lines but hide axis lines.</p></td>
    </tr>
    <tr>
      <th>headerLevel</th>
      <td><p><code>scalar integer</code> (<code>1</code>-<code>7</code>). R Markdown header level.</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code> or <code>NULL</code>. Character vector of
interesting groups. Must be formatted in camel case and intersect with
<code><a href='sampleData.html'>sampleData()</a></code> colnames.</p></td>
    </tr>
    <tr>
      <th>label</th>
      <td><p><code>boolean</code>. Overlay a cluster identitiy label on the plot.</p></td>
    </tr>
    <tr>
      <th>labelSize</th>
      <td><p><code>scalar integer</code>. Size of the text label.</p></td>
    </tr>
    <tr>
      <th>legend</th>
      <td><p><code>boolean</code>. Include plot legend.</p></td>
    </tr>
    <tr>
      <th>min</th>
      <td><p><code>scalar numeric</code>. Recommended minimum value cutoff.</p></td>
    </tr>
    <tr>
      <th>max</th>
      <td><p><code>scalar numeric</code>. Recommended maximum value cutoff.</p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>prefilter</th>
      <td><p><code>boolean</code>. Apply prefiltering to remove zero count genes.</p></td>
    </tr>
    <tr>
      <th>pipeline</th>
      <td><p><code>string</code>. Pipeline used to generate the samples.</p></td>
    </tr>
    <tr>
      <th>pointAlpha</th>
      <td><p><code>scalar numeric</code> (<code>0</code>-<code>1</code>). Alpha transparency level.
Useful when there many cells in the dataset, and some cells can be masked.</p></td>
    </tr>
    <tr>
      <th>pointsAsNumbers</th>
      <td><p><code>boolean</code>. Plot the points as numbers (<code>TRUE</code>) or
dots (<code>FALSE</code>).</p></td>
    </tr>
    <tr>
      <th>pointSize</th>
      <td><p><code>scalar numeric</code>. Cell point size.</p></td>
    </tr>
    <tr>
      <th>reduction</th>
      <td><p><code>string</code>. Dimensional reduction method to apply. Defaults to
t-SNE ("<code>TSNE</code>") but UMAP is also supported ("<code>UMAP</code>").</p></td>
    </tr>
    <tr>
      <th>return</th>
      <td><p><code>string</code>. Return type. Uses <code><a href='http://www.rdocumentation.org/packages/base/topics/match.arg'>base::match.arg()</a></code> internally and
defaults to the first argument in the <code>character</code> vector.</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p><code>string</code> or <code>NULL</code>. Plot title.</p></td>
    </tr>
    <tr>
      <th>trans</th>
      <td><p><code>string</code>. Name of the axis scale transformation to apply. See
<code>help("scale_x_continuous", "ggplot2")</code> for more information.</p></td>
    </tr>
    <tr>
      <th>trendline</th>
      <td><p><code>boolean</code>. Include trendline on plot.</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>Object to assign.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>Primary object.</p></td>
    </tr>
    <tr>
      <th>y</th>
      <td><p>Secondary object.</p></td>
    </tr>
    <tr>
      <th>zinbwave</th>
      <td><p><code>boolean</code>. Run <code><a href='http://www.rdocumentation.org/packages/zinbwave/topics/zinbwave'>zinbwave::zinbwave()</a></code> to automatically apply
a ZINB regression model to calculate <code>normalizedValues</code> and <code>weights</code>
matrices to by used for differential expression with DESeq2 or edgeR.
Note that this calculation should only be performed on <strong>filtered data</strong>.
For large datasets this can take a long time and use a lot of memory, so
this calculation is disabled by default.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>No value.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

