<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Known Markers Detected — knownMarkersDetected • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioSingleCell</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Known Markers Detected</h1>
    </div>

    
    <p>Known Markers Detected</p>
    

    <pre class="usage"><span class='fu'>knownMarkersDetected</span>(<span class='no'>all</span>, <span class='no'>known</span>, <span class='no'>...</span>)

<span class='co'># S4 method for grouped_df,tbl_df</span>
<span class='fu'>knownMarkersDetected</span>(<span class='no'>all</span>, <span class='no'>known</span>, <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='kw'>filterPromiscuous</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>promiscuousCutoff</span> <span class='kw'>=</span> <span class='fl'>5L</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>all</th>
      <td><p>All markers, generated by <code><a href='http://www.rdocumentation.org/packages/Seurat/topics/FindAllMarkers'>Seurat::FindAllMarkers()</a></code> return.</p></td>
    </tr>
    <tr>
      <th>known</th>
      <td><p>Known markers data.frame imported by
<code>readCellTypeMarkersFile()</code> or pulled from internal <a href='cellCycleMarkers.html'>cellCycleMarkers</a>
data.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>Alpha cutoff (adjusted P value; false discovery rate).</p></td>
    </tr>
    <tr>
      <th>filterPromiscuous</th>
      <td><p>Remove markers with poor specificity, that are
present in as many clusters as defined by <code>promiscuousCutoff</code>.</p></td>
    </tr>
    <tr>
      <th>promiscuousCutoff</th>
      <td><p>Minimum number of clusters required to consider a
gene marker promiscuous.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Grouped tibble.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Both the <code>all</code> and <code>known</code> objects must contain Ensembl gene
identifiers in the <code>ensgene</code> column. We must avoid any matching operations
based on the gene symbols, since these change often and can mismatch
easily.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p>Other Clustering Utilities: <code>plotKnownMarkersDetected</code>,
  <code>plotMarkers</code>, <code>plotPCElbow</code>,
  <code>plotTopMarkers</code>, <code>topBarcodes</code>,
  <code>topMarkers</code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>load</span>(<span class='fu'>system.file</span>(<span class='st'>"extdata/seuratAllMarkers.rda"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioSingleCell"</span>))

<span class='co'># Use internal known markers for Homo sapiens</span>
<span class='no'>knownMarkers</span> <span class='kw'>&lt;-</span> <span class='no'>cellTypeMarkers</span><span class='kw'>[[</span><span class='st'>"homoSapiens"</span>]]

<span class='co'># grouped_df</span>
<span class='fu'>knownMarkersDetected</span>(
    <span class='kw'>all</span> <span class='kw'>=</span> <span class='no'>seuratAllMarkers</span>,
    <span class='kw'>known</span> <span class='kw'>=</span> <span class='no'>knownMarkers</span>) <span class='kw'>%&gt;%</span>
    <span class='fu'><a href='http://www.rdocumentation.org/packages/tibble/topics/glimpse'>glimpse</a></span>()</div><div class='output co'>#&gt; Observations: 2
#&gt; Variables: 18
#&gt; $ cell           &lt;chr&gt; "Natural Killer Cell", "Natural Killer Cell"
#&gt; $ cluster        &lt;fct&gt; 1, 2
#&gt; $ ensgene        &lt;chr&gt; "ENSG00000149294", "ENSG00000149294"
#&gt; $ symbol         &lt;chr&gt; "NCAM1", "NCAM1"
#&gt; $ pct1           &lt;dbl&gt; 0.613, 0.848
#&gt; $ pct2           &lt;dbl&gt; 0.862, 0.787
#&gt; $ avgLogFC       &lt;dbl&gt; -1.0176323, 0.5827054
#&gt; $ pvalue         &lt;dbl&gt; 3.611922e-11, 7.177981e-06
#&gt; $ padj           &lt;dbl&gt; 3.611922e-08, 7.177981e-03
#&gt; $ description    &lt;chr&gt; "neural cell adhesion molecule 1 [Source:HGNC Symbol...
#&gt; $ biotype        &lt;chr&gt; "protein_coding", "protein_coding"
#&gt; $ broadClass     &lt;chr&gt; "coding", "coding"
#&gt; $ geneSeqStart   &lt;int&gt; 112961247, 112961247
#&gt; $ geneSeqEnd     &lt;int&gt; 113278436, 113278436
#&gt; $ seqName        &lt;chr&gt; "11", "11"
#&gt; $ seqStrand      &lt;int&gt; 1, 1
#&gt; $ seqCoordSystem &lt;chr&gt; "chromosome", "chromosome"
#&gt; $ entrez         &lt;list&gt; [4684, 4684]</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
