<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Load bcbio Single-Cell RNA-Seq Data — loadSingleCell • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioSingleCell</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load bcbio Single-Cell RNA-Seq Data</h1>
    </div>

    
    <p>Load bcbio Single-Cell RNA-Seq Data</p>
    

    <pre class="usage"><span class='fu'>loadSingleCell</span>(<span class='no'>uploadDir</span>, <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>, <span class='kw'>gtfFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>annotable</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ensemblVersion</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>prefilter</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uploadDir</th>
      <td><p>Path to final upload directory. This path is set when
running <code>bcbio_nextgen -w template</code>.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p>Sample barcode metadata file. Optional for runs
with demultiplixed index barcodes (e.g. SureCell), but otherwise required
for runs with multipliexed FASTQs containing multiple index barcodes (e.g.
inDrop). Consult the GitHub repo for examples and additional information.</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p>Character vector of interesting groups. First entry
is used for plot colors during quality control (QC) analysis. Entire vector
is used for PCA and heatmap QC functions.</p></td>
    </tr>
    <tr>
      <th>gtfFile</th>
      <td><p><em>Optional but recommended</em>. GTF (Gene Transfer Format) file,
which will be used for gene-to-symbol (<code>gene2symbol</code>) and
transcript-to-gene (<code>tx2gene</code>) annotation mappings.</p></td>
    </tr>
    <tr>
      <th>annotable</th>
      <td><p>User-defined gene annotations (a.k.a.
"annotable"), which will be slotted into <code>rowData()</code>. Typically this should
be left set to <code>TRUE</code>. By default, the function will automatically generate
an annotable from the annotations available on Ensembl. If set <code>FALSE</code>,
then <code>rowData()</code> inside the resulting <a href='bcbioSingleCell-class.html'>bcbioSingleCell</a> object will be left
empty. This is recommended for projects dealing with genes or transcripts
that are poorly annotated. Additionally, a manually constructed annotable
can be passed in as a data.frame, but this isn't generally recommended.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><em>Optional</em>. Organism name. Use the full latin name (e.g.
"Homo sapiens"), since this will be input downstream to
AnnotationHub/ensembldb. If set, this genome must be supported on Ensembl.
Normally this can be left <code>NULL</code>, and the function will attempt to detect
the organism automatically using <code><a href='detectOrganism.html'>detectOrganism()</a></code>.</p></td>
    </tr>
    <tr>
      <th>ensemblVersion</th>
      <td><p><em>Optional</em>. Ensembl release version. If <code>NULL</code>,
defaults to current release, and does not typically need to be
user-defined. This parameter can be useful for matching Ensembl annotations
against an outdated bcbio annotation build.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><em>Optional</em>. Genome build. Normally this can be left <code>NULL</code>
and the build will be detected from the bcbio run data. This can be set
manually (e.g. "hg19" for the older <em>Homo sapiens</em> reference genome). Note
that this must match the genome build identifier on Ensembl for annotations
to download correctly.</p></td>
    </tr>
    <tr>
      <th>prefilter</th>
      <td><p>Prefilter counts prior to quality control analysis.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments, to be stashed in the <code>metadata()</code> slot.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><a href='bcbioSingleCell-class.html'>bcbioSingleCell</a>.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>When working in RStudio, we recommend connecting to the bcbio-nextgen
run directory as a remote connection over
<a href='https://github.com/osxfuse/osxfuse/wiki/SSHFS'>sshfs</a>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Homo sapiens</span>
<span class='co'># Minimal working example dataset</span>
<span class='no'>extdataDir</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioSingleCell"</span>)
<span class='no'>uploadDir</span> <span class='kw'>&lt;-</span> <span class='fu'>file.path</span>(<span class='no'>extdataDir</span>, <span class='st'>"harvard_indrop_v3"</span>)
<span class='no'>sampleMetadataFile</span> <span class='kw'>&lt;-</span> <span class='fu'>file.path</span>(<span class='no'>extdataDir</span>, <span class='st'>"harvard_indrop_v3.xlsx"</span>)
<span class='no'>bcb</span> <span class='kw'>&lt;-</span> <span class='fu'>loadSingleCell</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='no'>sampleMetadataFile</span>)</div><div class='output co'>#&gt; <span class='message'>2017-11-10_bcbio</span></div><div class='output co'>#&gt; <span class='message'>2 samples detected</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml</span></div><div class='output co'>#&gt; <span class='message'>Reading log files</span></div><div class='output co'>#&gt; <span class='warning'>Warning: bcbio-nextgen.log missing</span></div><div class='output co'>#&gt; <span class='warning'>Warning: bcbio-nextgen-commands.log missing</span></div><div class='output co'>#&gt; <span class='message'>Reading data and program versions</span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  genome = col_character(),</span>
#&gt; <span class='message'>  resource = col_character(),</span>
#&gt; <span class='message'>  version = col_character()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>Organism: Homo sapiens</span>
#&gt; <span class='message'>Genome build: hg38</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Assuming unknown UMI is harvard-indrop-v3 (default)</span></div><div class='output co'>#&gt; <span class='message'>Reading harvard_indrop_v3.xlsx</span></div><div class='output co'>#&gt; <span class='message'>Loading Ensembl annotations from AnnotationHub</span>
#&gt; <span class='message'>2017-10-27</span></div><div class='output co'>#&gt; <span class='message'>EnsDB AH57757: Homo sapiens Ensembl 90</span></div><div class='output co'>#&gt; <span class='message'>Reading cellular barcode distributions</span></div><div class='output co'>#&gt;    |                                                  | 0 % ~calculating  |==                                                                      |   4%|=====                                                                   |   8%|========                                                                |  11%|===========                                                             |  15%|==============                                                          |  19%|=================                                                       |  23%|===================                                                     |  27%|======================                                                  |  31%|=========================                                               |  34%|========================                                        |  38%    1 MB|===========================                                     |  42%    1 MB|=============================                                   |  45%    1 MB|================================                                |  49%    1 MB|==================================                              |  53%    1 MB|====================================                            |  56%    1 MB|=======================================                         |  60%    1 MB|=========================================                       |  63%    1 MB|===========================================                     |  67%    1 MB|=============================================                   |  70%    1 MB|================================================                |  74%    2 MB|==================================================              |  77%    2 MB|====================================================            |  81%    2 MB|=======================================================         |  84%    2 MB|=========================================================       |  88%    2 MB|===========================================================     |  91%    2 MB|=============================================================   |  95%    2 MB|================================================================|  98%    2 MB|=================================================================| 100%    2 MB
#&gt;    |+++++++++++++++++++++++++                         | 50% ~00s             |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 00s</div><div class='output co'>#&gt; <span class='message'>Reading counts at gene level</span></div><div class='output co'>#&gt;    |                                                  | 0 % ~calculating  </div><div class='output co'>#&gt; <span class='message'>run1_AGAGGATA</span></div><div class='output co'>#&gt;    |+++++++++++++++++++++++++                         | 50% ~06s          </div><div class='output co'>#&gt; <span class='message'>run2_AGAGGATA</span></div><div class='output co'>#&gt;    |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 10s</div><div class='output co'>#&gt; <span class='error'>Error in (function (classes, fdef, mtable) {    methods &lt;- .findInheritedMethods(classes, fdef, mtable)    if (length(methods) == 1L)         return(methods[[1L]])    else if (length(methods) == 0L) {        cnames &lt;- paste0("\"", vapply(classes, as.character,             ""), "\"", collapse = ", ")        stop(gettextf("unable to find an inherited method for function %s for signature %s",             sQuote(fdef@generic), sQuote(cnames)), domain = NA)    }    else stop("Internal error in finding inherited methods; didn't return a unique method",         domain = NA)})(list(structure("dgCMatrix", package = "Matrix")), structure(function (object,     ...) {    standardGeneric("calculateMetrics")}, generic = structure("calculateMetrics", package = "bcbioSingleCell"), package = "bcbioSingleCell", group = list(), valueClass = character(0), signature = "object", default = `\001NULL\001`, skeleton = (function (object,     ...) stop("invalid call in method dispatch to 'calculateMetrics' (no default method)",     domain = NA))(object, ...), class = structure("nonstandardGenericFunction", package = "methods")),     &lt;environment&gt;): unable to find an inherited method for function 'calculateMetrics' for signature '"dgCMatrix"'</span></div><div class='input'>
# Mus musculus
# Run with Ensembl 88 transcriptome FASTA and GTF files
</div><span class='co'># NOT RUN {</span>
<span class='no'>bcb</span> <span class='kw'>&lt;-</span> <span class='fu'>loadSingleCell</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='st'>"indrop_rnaseq"</span>, <span class='st'>"final"</span>),
    <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"genotype"</span>, <span class='st'>"treatment"</span>),
    <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='st'>"meta"</span>, <span class='st'>"sample_metadata.xlsx"</span>),
    <span class='kw'>gtfFile</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(
        <span class='st'>"annotations"</span>,
        <span class='st'>"Mus_musculus.GRCm38.88.chr_patch_hapl_scaff.gtf.gz"</span>),
    <span class='kw'>ensemblVersion</span> <span class='kw'>=</span> <span class='fl'>88</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Rory Kirchner

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
