<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Load 10X Genomics CellRanger Data — loadCellRanger • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioSingleCell</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load 10X Genomics CellRanger Data</h1>
    </div>

    
    <p>Read <a href='https://www.10xgenomics.com/software/'>10x Genomics Chromium</a> cell
counts from <code>barcodes.tsv</code>, <code>genes.tsv</code>, and <code>matrix.mtx</code> files.</p>
    

    <pre class="usage"><span class='fu'>loadCellRanger</span>(<span class='no'>uploadDir</span>, <span class='no'>refdataDir</span>, <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>,
  <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>annotable</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>prefilter</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uploadDir</th>
      <td><p>Path to CellRanger output directory. This directory path
must contain <code>filtered_gene_bc_matrices*</code> as a child directory.</p></td>
    </tr>
    <tr>
      <th>refdataDir</th>
      <td><p>Directory path to cellranger reference annotation data.</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p>Character vector of interesting groups. First entry
is used for plot colors during quality control (QC) analysis. Entire vector
is used for PCA and heatmap QC functions.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p>Sample barcode metadata file. Optional for runs
with demultiplixed index barcodes (e.g. SureCell), but otherwise required
for runs with multipliexed FASTQs containing multiple index barcodes (e.g.
inDrop). Consult the GitHub repo for examples and additional information.</p></td>
    </tr>
    <tr>
      <th>annotable</th>
      <td><p>User-defined gene annotations (a.k.a.
"annotable"), which will be slotted into <code>rowData()</code>. Typically this should
be left set to <code>TRUE</code>. By default, the function will automatically generate
an annotable from the annotations available on Ensembl. If set <code>FALSE</code>,
then <code>rowData()</code> inside the resulting <a href='bcbioSingleCell-class.html'>bcbioSingleCell</a> object will be left
empty. This is recommended for projects dealing with genes or transcripts
that are poorly annotated. Additionally, a manually constructed annotable
can be passed in as a data.frame, but this isn't generally recommended.</p></td>
    </tr>
    <tr>
      <th>prefilter</th>
      <td><p>Prefilter counts prior to quality control analysis.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments, to be stashed in the <code>metadata()</code> slot.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><a href='bcbioSingleCell-class.html'>bcbioSingleCell</a>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function is a simplified version of <code><a href='loadSingleCell.html'>loadSingleCell()</a></code>
optimized for handling CellRanger output.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Unlike <code><a href='loadSingleCell.html'>loadSingleCell()</a></code>, the <code>organism</code>, <code>ensemblVersion</code>, and
<code>genomeBuild</code> are always detected automatically, based on the <code>refdataDir</code>
YAML metadata. Therefore, these parameters cannot be set by the user.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>extdataDir</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioSingleCell"</span>)
<span class='no'>uploadDir</span> <span class='kw'>&lt;-</span> <span class='fu'>file.path</span>(<span class='no'>extdataDir</span>, <span class='st'>"cellranger"</span>)
<span class='no'>refdataDir</span> <span class='kw'>&lt;-</span> <span class='fu'>file.path</span>(<span class='no'>extdataDir</span>, <span class='st'>"refdata-cellranger-hg19-1.2.0"</span>)
<span class='no'>sampleMetadataFile</span> <span class='kw'>&lt;-</span> <span class='fu'>file.path</span>(<span class='no'>extdataDir</span>, <span class='st'>"cellranger.csv"</span>)
<span class='no'>bcb</span> <span class='kw'>&lt;-</span> <span class='fu'>loadCellRanger</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>refdataDir</span> <span class='kw'>=</span> <span class='no'>refdataDir</span>,
    <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='no'>sampleMetadataFile</span>)</div><div class='output co'>#&gt; <span class='message'>CellRanger output directory structure:</span>
#&gt; <span class='message'>&lt;uploadDir&gt;/&lt;sampleName&gt;/outs/filtered_gene_bc_matrices*/&lt;genomeBuild&gt;/matrix.mtx</span></div><div class='output co'>#&gt; <span class='message'>1 samples detected</span></div><div class='output co'>#&gt; <span class='message'>Reading cellranger.csv</span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  fileName = col_character(),</span>
#&gt; <span class='message'>  description = col_character(),</span>
#&gt; <span class='message'>  sampleName = col_character(),</span>
#&gt; <span class='message'>  index = col_integer()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>Organism: Homo sapiens</span>
#&gt; <span class='message'>Genome build: hg19</span>
#&gt; <span class='message'>Ensembl release: 82</span></div><div class='output co'>#&gt; <span class='message'>Minimal working example doesn't contain a GTF file</span></div><div class='output co'>#&gt; <span class='message'>Obtaining gene2symbol from Ensembl</span></div><div class='output co'>#&gt; <span class='message'>Reading counts at gene level</span></div><div class='output co'>#&gt; <span class='message'>Calculating barcode metrics</span></div><div class='output co'>#&gt; <span class='message'>500 cellular barcodes detected</span></div><div class='output co'>#&gt; <span class='message'>11 mitochondrial genes detected</span></div><div class='output co'>#&gt; <span class='message'>500 cellular barcodes passed pre-filtering (100%)</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>bcb</span>)</div><div class='output co'>#&gt; class: bcbioSingleCell 
#&gt; dim: 500 500 
#&gt; metadata(23): version pipeline ... utilsSessionInfo devtoolsSessionInfo
#&gt; assays(1): assay
#&gt; rownames(500): ENSG00000005022 ENSG00000008018 ... ENSG00000265681
#&gt;   ENSG00000269893
#&gt; rowData names(10): ensgene symbol ... strand entrez
#&gt; colnames(500): aggregation_AAACCTGGTTTACTCT_1
#&gt;   aggregation_AAACGGGGTATCTGCA_1 ... aggregation_TTTGTCACACAACGTT_2
#&gt;   aggregation_TTTGTCACACTCGACG_2
#&gt; colData names(6): nUMI nGene ... log10GenesPerUMI mitoRatio</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
