<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read 10X Genomics Cell Ranger Data — readCellRanger • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Read 10X Genomics Cell Ranger Data — readCellRanger" />

<meta property="og:description" content="Read 10x Genomics Chromium cell
counts from barcodes.tsv, genes.tsv, and matrix.mtx files." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read 10X Genomics Cell Ranger Data</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/readCellRanger.R'><code>R/readCellRanger.R</code></a></small>
    <div class="hidden name"><code>readCellRanger.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Read <a href='https://www.10xgenomics.com/software/'>10x Genomics Chromium</a> cell
counts from <code>barcodes.tsv</code>, <code>genes.tsv</code>, and <code>matrix.mtx</code> files.</p>
    
    </div>

    <pre class="usage"><span class='fu'>readCellRanger</span>(<span class='no'>uploadDir</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"mtx"</span>, <span class='st'>"hdf5"</span>), <span class='kw'>filtered</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>refdataDir</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>, <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>spikeNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uploadDir</th>
      <td><p><code>string</code>. Path to Cell Ranger output directory. This
directory path must contain <code>filtered_gene_bc_matrices*</code> as a child
directory.</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p><code>string</code>. Output format, either MatrixMarket ("<code>mtx</code>") or HDF5
("<code>hdf5</code>").</p></td>
    </tr>
    <tr>
      <th>filtered</th>
      <td><p><code>boolean</code>. Use filtered (recommended) or raw counts. Note
that raw counts still contain only whitelisted cellular barcodes.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code>string</code> or <code>NULL</code>. Organism name. Use the full Latin name
(e.g. "Homo sapiens"), since this will be input downstream to AnnotationHub
and ensembldb, unless <code>gffFile</code> is set. If left <code>NULL</code> (<em>not recommended</em>),
the function call will skip loading gene-level annotations into
<code>rowRanges()</code>. This can be useful for poorly annotation genomes or
experiments involving multiple genomes.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p><code>string</code> or <code>NULL</code>. Sample barcode metadata file.
Optional for runs with demultiplixed index barcodes (e.g. SureCell) and can
be left <code>NULL</code>, but otherwise required for runs with multipliexed FASTQs
containing multiple index barcodes (e.g. inDrop).</p></td>
    </tr>
    <tr>
      <th>refdataDir</th>
      <td><p><code>string</code> or <code>NULL</code>. Directory path to Cell Ranger reference
annotation data.</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code> or <code>NULL</code>. Character vector of
interesting groups. Must be formatted in camel case and intersect with
<code><a href='sampleData.html'>sampleData()</a></code> colnames.</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code> or <code>NULL</code>. Vector indicating which
<code>assay()</code> rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code> or <code>NULL</code>. Vector indicating which <code>assay()</code>
rows denote spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments, to be stashed in the <code>metadata()</code> slot.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>SingleCellExperiment</code>.</p>
    
    <h2 class="hasAnchor" id="directory-structure"><a class="anchor" href="#directory-structure"></a>Directory structure</h2>

    
    <p>Cell Ranger can vary in its output directory structure, but we're requiring a
single, consistent directory structure for all datasets, even those that only
contain a single sample:</p>
<pre>
file.path(
    "&lt;uploadDir&gt;",
    "&lt;sampleName&gt;",
    "outs",
    "filtered_gene_bc_matrices*",
    "outs",
    "&lt;genomeBuild&gt;",
    "matrix.mtx"
)
</pre>
    
    <h2 class="hasAnchor" id="sample-metadata"><a class="anchor" href="#sample-metadata"></a>Sample metadata</h2>

    
    <p>A user-supplied sample metadata file defined by <code>sampleMetadataFile</code> is
required for multiplexed datasets. Otherwise this can be left <code>NULL</code>, and
minimal sample data will be used, based on the directory names.</p>
    
    <h2 class="hasAnchor" id="reference-data"><a class="anchor" href="#reference-data"></a>Reference data</h2>

    
    <p>We strongly recommend supplying the corresponding reference data required for
Cell Ranger with the <code>refdataDir</code> argument. It will convert the gene
annotations defined in the GTF file into a <code>GRanges</code> object, which get
slotted in <code>rowRanges()</code>. Otherwise, the function will attempt to use the
most current annotations available from Ensembl, and some gene IDs may not
match, due to deprecation in the current Ensembl release.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>uploadDir</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata/cellranger"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioSingleCell"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>readCellRanger</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Homo sapiens"</span>
)</div><div class='output co'>#&gt; <span class='message'>1 sample(s) detected</span></div><div class='output co'>#&gt; <span class='message'>Reading counts at gene level</span></div><div class='output co'>#&gt; <span class='message'>Reading counts in mtx format</span></div><div class='output co'>#&gt; <span class='message'>/Users/mike/GitHub/hbc/bcbioSingleCell/inst/extdata/cellranger/pbmc/outs/filtered_gene_bc_matrices</span></div><div class='output co'>#&gt; <span class='message'>Multiplexed (aggregated) samples detected</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.12.0 (2018-04-30)</span></div><div class='output co'>#&gt; <span class='message'>Ensembl 92 EnsDb for Homo Sapiens</span></div><div class='output co'>#&gt; <span class='message'>id: "AH60977"</span>
#&gt; <span class='message'>organism: "Homo sapiens"</span>
#&gt; <span class='message'>build: "GRCh38"</span>
#&gt; <span class='message'>release: "92"</span>
#&gt; <span class='message'>format: "genes"</span></div><div class='output co'>#&gt; <span class='message'>Standardizing the metadata columns</span></div><div class='output co'>#&gt; <span class='message'>Defining broad class using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Sorting ranges by geneID</span></div><div class='output co'>#&gt; <span class='message'>Calculating cellular barcode metrics</span></div><div class='output co'>#&gt; <span class='message'>500 cells detected</span></div><div class='output co'>#&gt; <span class='message'>312 coding genes</span></div><div class='output co'>#&gt; <span class='message'>0 mitochondrial genes</span></div><div class='output co'>#&gt; <span class='message'>500 / 500 cellular barcodes passed pre-filtering (100%)</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Unannotated rows detected (9):</span>
#&gt; <span class='warning'>ENSG00000236743, ENSG00000278566, ENSG00000273547, ENSG00000279244, ENSG00000...</span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: SingleCellExperiment 
#&gt; dim: 500 500 
#&gt; metadata(20): version pipeline ... utilsSessionInfo devtoolsSessionInfo
#&gt; assays(1): counts
#&gt; rownames(500): ENSG00000243485 ENSG00000237613 ... ENSG00000230703
#&gt;   ENSG00000142677
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(500): pbmc_1_AAACCTGAGAAGGCCT pbmc_1_AAACCTGAGACAGACC ...
#&gt;   pbmc_1_ACTTACTCACTGCCAG pbmc_1_ACTTACTCAGTACACT
#&gt; colData names(10): sampleID nUMI ... description index
#&gt; reducedDimNames(0):
#&gt; spikeNames(0):</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#directory-structure">Directory structure</a></li>

      <li><a href="#sample-metadata">Sample metadata</a></li>

      <li><a href="#reference-data">Reference data</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

