<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Sanitize Seurat Markers — sanitizeSeuratMarkers • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Sanitize Seurat Markers — sanitizeSeuratMarkers" />

<meta property="og:description" content="Currently only Seurat markers are supported." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Sanitize Seurat Markers</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/sanitizeSeuratMarkers.R'><code>R/sanitizeSeuratMarkers.R</code></a></small>
    <div class="hidden name"><code>sanitizeSeuratMarkers.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Currently only Seurat markers are supported.</p>
    
    </div>

    <pre class="usage"><span class='fu'>sanitizeSeuratMarkers</span>(<span class='no'>data</span>, <span class='no'>rowRanges</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p><code>data.frame</code>. <code><a href='http://www.rdocumentation.org/packages/Seurat/topics/FindAllMarkers'>Seurat::FindAllMarkers()</a></code> or
<code><a href='http://www.rdocumentation.org/packages/Seurat/topics/FindMarkers'>Seurat::FindMarkers()</a></code> return.</p></td>
    </tr>
    <tr>
      <th>rowRanges</th>
      <td><p><code>GRanges</code>. Gene annotations. Must contain <code>geneName</code> column
that corresponds to the gene names in the <code>seurat</code> object used to generate
the markers <code>data.frame</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>data.frame</code>, arranged by adjusted P value.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><code><a href='http://www.rdocumentation.org/packages/Seurat/topics/FindAllMarkers'>Seurat::FindAllMarkers()</a></code> maps the counts matrix rownames correctly in
the <code>gene</code> column, whereas <code><a href='http://www.rdocumentation.org/packages/Seurat/topics/FindMarkers'>Seurat::FindMarkers()</a></code> maps them correctly in
the rownames of the returned marker <code>data.frame</code>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Clustering Functions: <code><a href='cellTypesPerCluster.html'>cellTypesPerCluster</a></code>,
  <code><a href='knownMarkersDetected.html'>knownMarkersDetected</a></code>,
  <code><a href='plotCellTypesPerCluster.html'>plotCellTypesPerCluster</a></code>,
  <code><a href='plotFeature.html'>plotFeatureTSNE</a></code>,
  <code><a href='plotMarker.html'>plotKnownMarkersDetected</a></code>,
  <code><a href='plotPCElbow.html'>plotPCElbow</a></code>, <code><a href='plotDimensionalReduction.html'>plotTSNE</a></code>,
  <code><a href='topMarkers.html'>topMarkers</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># seurat ====</span>
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>seurat_small</span>

<span class='co'># `FindAllMarkers()` return</span>
<span class='fu'>invisible</span>(<span class='fu'>capture.output</span>(
    <span class='no'>all_markers</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/Seurat/topics/FindAllMarkers'>FindAllMarkers</a></span>(<span class='no'>object</span>)
))
<span class='no'>all_sanitized</span> <span class='kw'>&lt;-</span> <span class='fu'>sanitizeSeuratMarkers</span>(
    <span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>all_markers</span>,
    <span class='kw'>rowRanges</span> <span class='kw'>=</span> <span class='fu'>rowRanges</span>(<span class='no'>object</span>)
)
<span class='fu'>glimpse</span>(<span class='no'>all_sanitized</span>)</div><div class='output co'>#&gt; Observations: 952
#&gt; Variables: 18
#&gt; $ cluster        &lt;fct&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...
#&gt; $ avgLogFC       &lt;dbl&gt; 1.9829602, 3.0536375, 2.8619805, 2.2483912, 1.804789...
#&gt; $ pct1           &lt;dbl&gt; 0.921, 1.000, 1.000, 0.872, 1.000, 0.988, 1.000, 1.0...
#&gt; $ pct2           &lt;dbl&gt; 0.137, 0.807, 0.890, 0.173, 0.866, 0.384, 0.446, 0.4...
#&gt; $ rowname        &lt;chr&gt; "VCAN", "S100A8", "S100A9", "S100A12", "LYZ", "AC020...
#&gt; $ pvalue         &lt;dbl&gt; 1.369269e-74, 8.446076e-72, 1.100817e-71, 8.711241e-...
#&gt; $ padj           &lt;dbl&gt; 6.846344e-72, 4.223038e-69, 5.504085e-69, 4.355621e-...
#&gt; $ seqnames       &lt;fct&gt; 5, 1, 1, 1, 12, 12, 3, 9, 22, 3, 1, 19, 22, X, 1, 22...
#&gt; $ start          &lt;int&gt; 83471465, 153390032, 153357854, 153373706, 69348341,...
#&gt; $ end            &lt;int&gt; 83582303, 153391188, 153361027, 153375649, 69354234,...
#&gt; $ width          &lt;int&gt; 110839, 1157, 3174, 1944, 5894, 733, 16729, 12074, 1...
#&gt; $ strand         &lt;fct&gt; +, -, +, -, +, -, +, -, -, -, +, +, +, -, -, +, -, -...
#&gt; $ broadClass     &lt;fct&gt; coding, coding, coding, coding, coding, other, codin...
#&gt; $ description    &lt;fct&gt; versican [Source:HGNC Symbol;Acc:HGNC:2464], S100 ca...
#&gt; $ geneBiotype    &lt;fct&gt; protein_coding, protein_coding, protein_coding, prot...
#&gt; $ geneID         &lt;chr&gt; "ENSG00000038427", "ENSG00000143546", "ENSG000001632...
#&gt; $ geneName       &lt;fct&gt; VCAN, S100A8, S100A9, S100A12, LYZ, AC020656.1, CSTA...
#&gt; $ seqCoordSystem &lt;fct&gt; chromosome, chromosome, chromosome, chromosome, chro...</div><div class='input'>
<span class='co'># `FindMarkers()` return</span>
<span class='fu'>invisible</span>(<span class='fu'>capture.output</span>(
    <span class='no'>ident_3_markers</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/Seurat/topics/FindMarkers'>FindMarkers</a></span>(
        <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
        <span class='kw'>ident.1</span> <span class='kw'>=</span> <span class='st'>"3"</span>,
        <span class='kw'>ident.2</span> <span class='kw'>=</span> <span class='kw'>NULL</span>
    )
))
<span class='no'>ident_3_sanitized</span> <span class='kw'>&lt;-</span> <span class='fu'>sanitizeSeuratMarkers</span>(
    <span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>ident_3_markers</span>,
    <span class='kw'>rowRanges</span> <span class='kw'>=</span> <span class='fu'>rowRanges</span>(<span class='no'>object</span>)
)
<span class='fu'>glimpse</span>(<span class='no'>ident_3_sanitized</span>)</div><div class='output co'>#&gt; Observations: 181
#&gt; Variables: 16
#&gt; $ avgLogFC       &lt;dbl&gt; 2.0535014, 2.0801853, 3.0662657, -1.3856482, -0.6757...
#&gt; $ pct1           &lt;dbl&gt; 0.900, 0.743, 0.843, 0.657, 1.000, 0.943, 0.357, 0.5...
#&gt; $ pct2           &lt;dbl&gt; 0.153, 0.191, 0.351, 0.965, 1.000, 0.993, 0.881, 0.9...
#&gt; $ pvalue         &lt;dbl&gt; 1.534176e-50, 4.789948e-31, 1.696343e-29, 4.771235e-...
#&gt; $ padj           &lt;dbl&gt; 7.670879e-48, 2.394974e-28, 8.481717e-27, 2.385617e-...
#&gt; $ seqnames       &lt;fct&gt; 14, 19, 2, 1, X, 1, 9, 22, 1, 21, 1, 3, 21, 12, 6, 1...
#&gt; $ start          &lt;int&gt; 105851708, 41877120, 88857161, 151982915, 12975108, ...
#&gt; $ end            &lt;int&gt; 105856218, 41881372, 88857683, 151994390, 12977227, ...
#&gt; $ width          &lt;int&gt; 4511, 4253, 523, 11476, 2120, 6524, 18637, 11729, 15...
#&gt; $ strand         &lt;fct&gt; -, +, -, -, +, -, +, +, -, -, -, +, -, +, +, +, +, -...
#&gt; $ broadClass     &lt;fct&gt; ig, coding, ig, coding, coding, coding, coding, codi...
#&gt; $ description    &lt;fct&gt; immunoglobulin heavy constant mu [Source:HGNC Symbol...
#&gt; $ geneBiotype    &lt;fct&gt; IG_C_gene, protein_coding, IG_C_gene, protein_coding...
#&gt; $ geneID         &lt;chr&gt; "ENSG00000211899", "ENSG00000105369", "ENSG000002115...
#&gt; $ geneName       &lt;fct&gt; IGHM, CD79A, IGKC, S100A10, TMSB4X, S100A4, ANXA1, T...
#&gt; $ seqCoordSystem &lt;fct&gt; chromosome, chromosome, chromosome, chromosome, chro...</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

