<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Update an Object to Its Current Class Definition — updateObject • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Update an Object to Its Current Class Definition — updateObject" />

<meta property="og:description" content="updateObject is a generic function that returns an instance
  of object updated to its current class definition." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Update an Object to Its Current Class Definition</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/updateObject-methods.R'><code>R/updateObject-methods.R</code></a></small>
    <div class="hidden name"><code>updateObject.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p><code>updateObject</code> is a generic function that returns an instance
  of <code>object</code> updated to its current class definition.</p>
    
    </div>

    <pre class="usage"><span class='co'># S4 method for bcbioSingleCell</span>
<span class='fu'>updateObject</span>(<span class='no'>object</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object to be updated for <code>updateObject</code> and
    <code>updateObjectFromSlots</code>.</p>
<p>Object for slot information to be extracted from for
    <code>getObjectSlots</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Valid object.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Updating objects is primarily useful when an object has been
  serialized (e.g., stored to disk) for some time (e.g., months), and the
  class definition has in the mean time changed. Because of the changed
  class definition, the serialized instance is no longer valid.</p>
<p><code>updateObject</code> requires that the class of the returned object be
  the same as the class of the argument <code>object</code>, and that the
  object is valid (see <code><a href='http://www.rdocumentation.org/packages/methods/topics/validObject'>validObject</a></code>). By default,
  <code>updateObject</code> has the following behaviors:</p><dl class='dl-horizontal'>
    <dt><code>updateObject(ANY, &#8230;, verbose=FALSE)</code></dt><dd><p>By default, <code>updateObject</code> uses heuristic methods to determine
      whether the object should be the `new' S4 type (introduced in R 2.4.0),
      but is not. If the heuristics indicate an update is required,
      the <code>updateObjectFromSlots</code> function tries to update the
      object. The default method returns the original S4 object or the
      successfully updated object, or issues an error if an update is
      required but not possible.
      The optional named argument <code>verbose</code> causes a message to be
      printed describing the action.
      Arguments <code>&#8230;</code> are passed to <code>updateObjectFromSlots</code>.</p></dd>
    <dt><code>updateObject(list, &#8230;, verbose=FALSE)</code></dt><dd><p>Visit each element in <code>list</code>, applying
      <code>updateObject(list[[elt]], &#8230;, verbose=verbose)</code>.</p></dd>
    <dt><code>updateObject(environment, &#8230;, verbose=FALSE)</code></dt><dd><p>Visit each element in <code>environment</code>, applying
      <code>updateObject(environment[[elt]], &#8230;, verbose=verbose)</code></p></dd>
    <dt><code>updateObject(formula, &#8230;, verbose=FALSE)</code></dt><dd><p>Do nothing; the environment of the formula may be too general
      (e.g., <code>R_GlobalEnv</code>) to attempt an update.</p></dd>
    <dt><code>updateObject(envRefClass, &#8230;, verbose=FALSE)</code></dt><dd><p>Attempt to update objects from fields using a strategy like
      <code>updateObjectFromSlots</code> Method 1.</p></dd>
  </dl>
    <p><code>updateObjectFromSlots(object, objclass=class(object), &#8230;,
  verbose=FALSE)</code> is a utility function that identifies the intersection
  of slots defined in the <code>object</code> instance and <code>objclass</code>
  definition. Under Method 1, the corresponding elements in
  <code>object</code> are then updated (with <code>updateObject(elt, &#8230;,
  verbose=verbose)</code>) and used as arguments to a call to <code>new(class,
  &#8230;)</code>, with <code>&#8230;</code> replaced by slots from the original
  object. If this fails, then Method 2 tries <code>new(class)</code> and
  assigns slots of <code>object</code> to the newly created instance.</p>
<p><code>getObjectSlots(object)</code> extracts the slot names and contents from
  <code>object</code>. This is useful when <code>object</code> was created by a class
  definition that is no longer current, and hence the contents of
  <code>object</code> cannot be determined by accessing known slots.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other S4 Object: <code><a href='bcbioSingleCell.html'>bcbioSingleCell</a></code>,
  <code><a href='coerce.html'>coerce</a></code>, <code><a href='extract.html'>extract</a></code>,
  <code><a href='show.html'>show</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># bcbioSingleCell ====</span>
<span class='fu'>updateObject</span>(<span class='no'>indrops_small</span>)</div><div class='output co'>#&gt; <span class='message'>Upgrading from 0.1.18 to 0.1.18</span></div><div class='output co'>#&gt; bcbioSingleCell 0.1.18
#&gt; class: SingleCellExperiment 
#&gt; dim: 500 500 
#&gt; metadata(35): version pipeline ... filterSummary previousVersion
#&gt; assays(4): counts normalizedValues weights logcounts
#&gt; rownames(500): ENSG00000000457 ENSG00000002586 ... ENSG00000283060
#&gt;   ENSG00000283391
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(500): multiplexed_AAAAAAAA_AAACACTA_ATCCCACG
#&gt;   multiplexed_AAAAAAAA_AAACACTA_CTTAGGTA ...
#&gt;   multiplexed_AAAAAAAA_TGTTATCA_ACGCAGAG
#&gt;   multiplexed_AAAAAAAA_TGTTATCA_CAAGTCAT
#&gt; colData names(15): sampleID nCount ... aggregate revcomp
#&gt; reducedDimNames(0):
#&gt; spikeNames(0):
#&gt; ============================================================
#&gt; Upload Dir: "/Users/mike/GitHub/hbc/bcbioSingleCell/inst/extdata/indrops"
#&gt; Upload Date: 2018-01-01
#&gt; R Load Date: 2018-07-31
#&gt; Level: "genes"
#&gt; Organism: "Homo sapiens"
#&gt; Interesting Groups: "sampleName"
#&gt; Metadata File: "inst/extdata/indrops/metadata.csv"
#&gt; AnnotationHub: "AH57757"
#&gt; Ensembl Release: "90"
#&gt; Genome Build: "GRCh38"
#&gt; Filtered: TRUE
#&gt; zinbwave: TRUE</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

